{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import google.generativeai as genai\n",
    "import json\n",
    "import os\n",
    "import time\n",
    "from pathlib import Path\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "genai.configure(api_key=os.environ[\"GEMINI_API_KEY\"])\n",
    "\n",
    "# Create the model\n",
    "# See https://ai.google.dev/api/python/google/generativeai/GenerativeModel\n",
    "generation_config = {\n",
    "  \"temperature\": 0.3,\n",
    "  \"top_p\": 0.95,\n",
    "  \"top_k\": 64,\n",
    "  \"max_output_tokens\": 8192,\n",
    "  \"response_mime_type\": \"application/json\",\n",
    "}\n",
    "safety_settings = [\n",
    "  {\n",
    "    \"category\": \"HARM_CATEGORY_HARASSMENT\",\n",
    "    \"threshold\": \"BLOCK_MEDIUM_AND_ABOVE\",\n",
    "  },\n",
    "  {\n",
    "    \"category\": \"HARM_CATEGORY_HATE_SPEECH\",\n",
    "    \"threshold\": \"BLOCK_MEDIUM_AND_ABOVE\",\n",
    "  },\n",
    "  {\n",
    "    \"category\": \"HARM_CATEGORY_SEXUALLY_EXPLICIT\",\n",
    "    \"threshold\": \"BLOCK_MEDIUM_AND_ABOVE\",\n",
    "  },\n",
    "  {\n",
    "    \"category\": \"HARM_CATEGORY_DANGEROUS_CONTENT\",\n",
    "    \"threshold\": \"BLOCK_MEDIUM_AND_ABOVE\",\n",
    "  },\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = \"data/eutils_raw/\"\n",
    "api_docs = []\n",
    "\n",
    "for _file in Path(data_dir).iterdir():\n",
    "    with open(_file) as f:\n",
    "        doc = f.readlines()\n",
    "    doc = \"\".join(doc)\n",
    "    api_docs.append(doc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"data/prompts/apidoc2json.md\") as f:\n",
    "    base_prompt = f.readlines()\n",
    "base_prompt = \"\".join(base_prompt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 6/6 [00:52<00:00,  8.67s/it]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "model = genai.GenerativeModel(\n",
    "  model_name=\"gemini-1.5-flash-latest\",\n",
    "  safety_settings=safety_settings,\n",
    "  generation_config=generation_config,\n",
    "  system_instruction=\"You are a helpful, meticulous, scientific programming assistant. You carefully read API documentation and generate JSON.\",\n",
    ")\n",
    "\n",
    "\n",
    "responses = []\n",
    "for doc in tqdm(api_docs):\n",
    "  prompt = base_prompt.replace(\"{API_DOC}\", doc)\n",
    "\n",
    "  response = model.generate_content(prompt)\n",
    "  text = json.loads(response.text)\n",
    "  responses.append(text)\n",
    "  time.sleep(0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[{'user_name': 'template_usrname',\n",
       "   'api_name': 'EFetch',\n",
       "   'api_call': 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db={db}&id={id}&retmode={retmode}&rettype={rettype}',\n",
       "   'api_version': '1.0',\n",
       "   'api_arguments': {'db': 'string (The database to retrieve records from. Must be a valid Entrez database name. Default is pubmed)',\n",
       "    'id': 'string (A single UID or a comma-delimited list of UIDs to retrieve. All UIDs must be from the database specified by db)',\n",
       "    'retmode': 'string (The data format of the records returned. Options include plain text, HTML, or XML. See Table 1 in Chapter 2 for a full list of allowed values for each database)',\n",
       "    'rettype': 'string (The record view returned. Options include Abstract or MEDLINE from PubMed, or GenPept or FASTA from protein. See Table 1 in Chapter 2 for a full list of allowed values for each database)'},\n",
       "   'functionality': 'Retrieves formatted data records for a list of input UIDs',\n",
       "   'env_requirements': ['requests'],\n",
       "   'example_code': \"import requests\\n\\n# Fetch PMIDs 17284678 and 9997 as text abstracts\\napi_url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=pubmed&id=17284678,9997&retmode=text&rettype=abstract'\\nresponse = requests.get(api_url)\\nprint(response.text)\",\n",
       "   'meta_data': {'description': 'This API method allows users to retrieve formatted data records for a list of input UIDs from various Entrez databases.',\n",
       "    'documentation': 'https://www.ncbi.nlm.nih.gov/books/NBK25499/'},\n",
       "   'questions': ['How can I retrieve the full text of a PubMed article given its PMID?',\n",
       "    'What is the process for obtaining the FASTA sequence of a protein given its GI number?',\n",
       "    'I need to retrieve the GenBank flat file for a specific nucleotide sequence. What API call should I use?',\n",
       "    'I want to download the abstract of a PubMed article in XML format. What are the necessary parameters?',\n",
       "    'How can I retrieve the first 100 bases of a nucleotide sequence on the minus strand?']}],\n",
       " [{'user_name': 'template_usrname',\n",
       "   'api_name': 'EPost API',\n",
       "   'api_call': 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/epost.fcgi?db={db}&id={id}&WebEnv={WebEnv}',\n",
       "   'api_version': '1.0',\n",
       "   'api_arguments': {'db': \"string (The Entrez database containing the UIDs. Default is 'pubmed')\",\n",
       "    'id': 'string (A single UID or a comma-delimited list of UIDs. For PubMed, no more than 10,000 UIDs can be included in a single URL request. For other databases there is no set maximum for the number of UIDs that can be passed to epost, but if more than about 200 UIDs are to be posted, the request should be made using the HTTP POST method.)',\n",
       "    'WebEnv': 'string (The Web Environment that will receive the UID list sent by post. EPost will create a new query key associated with that Web Environment. Usually this WebEnv value is obtained from the output of a previous ESearch, EPost or ELink call. If no WebEnv parameter is provided, EPost will create a new Web Environment and post the UID list to query_key 1.)'},\n",
       "   'functionality': 'Uploads a list of UIDs to the Entrez History server',\n",
       "   'env_requirements': ['requests'],\n",
       "   'example_code': \"import requests\\n\\n# Post records to PubMed\\napi_url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/epost.fcgi?db=pubmed&id=11237011,12466850'\\nresponse = requests.get(api_url)\\nprint(response.text)\",\n",
       "   'meta_data': {'description': 'The EPost API allows users to upload a list of UIDs to the Entrez History server, append a list of UIDs to an existing set of UID lists attached to a Web Environment, and retrieve a Web Environment containing a list of UIDs.',\n",
       "    'documentation': 'https://www.ncbi.nlm.nih.gov/books/NBK25499/'},\n",
       "   'questions': ['I have a list of PubMed IDs. How can I upload them to the Entrez History server?',\n",
       "    'I want to add a list of protein IDs to an existing Web Environment. How can I do this?',\n",
       "    'How can I retrieve a Web Environment containing a list of UIDs?',\n",
       "    'I have a list of 10,000 PubMed IDs. How can I upload them to the Entrez History server?',\n",
       "    'I have a list of 200 protein IDs. How can I upload them to the Entrez History server?',\n",
       "    'I have a list of 500 nucleotide IDs. How can I upload them to the Entrez History server?']}],\n",
       " [{'user_name': 'template_usrname',\n",
       "   'api_name': 'ESummary API',\n",
       "   'api_call': 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db={db}&id={id}&retmode={retmode}&retstart={retstart}&retmax={retmax}&version={version}',\n",
       "   'api_version': '1.0',\n",
       "   'api_arguments': {'db': 'string (The database from which to retrieve DocSums. The value must be a valid Entrez database name (default = pubmed).)',\n",
       "    'id': 'string (UID list. Either a single UID or a comma-delimited list of UIDs may be provided. All of the UIDs must be from the database specified by db. There is no set maximum for the number of UIDs that can be passed to ESummary, but if more than about 200 UIDs are to be provided, the request should be made using the HTTP POST method.)',\n",
       "    'retmode': 'string (Retrieval type. Determines the format of the returned output. The default value is ‘xml’ for ESummary XML, but ‘json’ is also supported to return output in JSON format.)',\n",
       "    'retstart': 'integer (Sequential index of the first DocSum to be retrieved (default=1, corresponding to the first record of the entire set). This parameter can be used in conjunction with retmax to download an arbitrary subset of DocSums from the input set.)',\n",
       "    'retmax': 'integer (Total number of DocSums from the input set to be retrieved, up to a maximum of 10,000. If the total set is larger than this maximum, the value of retstart can be iterated while holding retmax constant, thereby downloading the entire set in batches of size retmax.)',\n",
       "    'version': 'string (Used to specify version 2.0 ESummary XML. The only supported value is ‘2.0’. When present, ESummary will return version 2.0 DocSum XML that is unique to each Entrez database and that often contains more data than the default DocSum XML.)'},\n",
       "   'functionality': 'Returns document summaries (DocSums) for a list of input UIDs',\n",
       "   'env_requirements': ['requests'],\n",
       "   'example_code': \"import requests\\n\\n# Get summaries for PubMed IDs 11850928 and 11482001\\napi_url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=pubmed&id=11850928,11482001'\\nresponse = requests.get(api_url)\\nprint(response.text)\",\n",
       "   'meta_data': {'description': 'The ESummary API is a powerful tool for retrieving document summaries (DocSums) from various Entrez databases. It allows you to fetch summaries for a list of input UIDs, providing valuable information about each document.',\n",
       "    'documentation': 'https://www.ncbi.nlm.nih.gov/books/NBK25499/'},\n",
       "   'questions': ['How can I retrieve summaries for a list of PubMed IDs?',\n",
       "    'What is the maximum number of UIDs that can be passed to ESummary in a single request?',\n",
       "    'How can I retrieve summaries in JSON format?',\n",
       "    'I need to retrieve summaries for a large set of UIDs. How can I download the entire set in batches?',\n",
       "    'How can I retrieve version 2.0 ESummary XML for a specific database?']}],\n",
       " [{'user_name': 'template_usrname',\n",
       "   'api_name': 'NCBI ESearch API',\n",
       "   'api_call': 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db={db}&term={term}&usehistory={usehistory}&WebEnv={WebEnv}&query_key={query_key}&retstart={retstart}&retmax={retmax}&rettype={rettype}&retmode={retmode}&sort={sort}&field={field}&idtype={idtype}&datetype={datetype}&reldate={reldate}&mindate={mindate}&maxdate={maxdate}',\n",
       "   'api_version': '1.0',\n",
       "   'api_arguments': {'db': 'string (The Entrez database to search. Default: pubmed)',\n",
       "    'term': \"string (Entrez text query. All special characters must be URL encoded. Spaces may be replaced by '+' signs.)\",\n",
       "    'usehistory': \"string (When set to 'y', ESearch will post the UIDs resulting from the search operation onto the History server. Default: 'n')\",\n",
       "    'WebEnv': 'string (Web environment string returned from a previous ESearch, EPost or ELink call. When provided, ESearch will post the results of the search operation to this pre-existing WebEnv, thereby appending the results to the existing environment.)',\n",
       "    'query_key': 'integer (Integer query key returned by a previous ESearch, EPost or ELink call. When provided, ESearch will find the intersection of the set specified by query_key and the set retrieved by the query in term (i.e. joins the two with AND).)',\n",
       "    'retstart': 'integer (Sequential index of the first UID in the retrieved set to be shown in the XML output. Default: 0)',\n",
       "    'retmax': 'integer (Total number of UIDs from the retrieved set to be shown in the XML output. Default: 20)',\n",
       "    'rettype': \"string (Retrieval type. Allowed values: 'uilist' (default), 'count'.)\",\n",
       "    'retmode': \"string (Retrieval type. Determines the format of the returned output. Allowed values: 'xml' (default), 'json'.\",\n",
       "    'sort': 'string (Specifies the method used to sort UIDs in the ESearch output. The available values vary by database (db) and may be found in the Display Settings menu on an Entrez search results page.)',\n",
       "    'field': 'string (Search field. If used, the entire search term will be limited to the specified Entrez field.)',\n",
       "    'idtype': \"string (Specifies the type of identifier to return for sequence databases (nuccore, popset, protein). Default: GI numbers. Allowed values: 'acc' (accession.version identifiers).)\",\n",
       "    'datetype': \"string (Type of date used to limit a search. The allowed values vary between Entrez databases, but common values are 'mdat' (modification date), 'pdat' (publication date) and 'edat' (Entrez date).)\",\n",
       "    'reldate': 'integer (When reldate is set to an integer n, the search returns only those items that have a date specified by datetype within the last n days.)',\n",
       "    'mindate': 'string (Date range used to limit a search result by the date specified by datetype. Format: YYYY/MM/DD. Must be used with maxdate.)',\n",
       "    'maxdate': 'string (Date range used to limit a search result by the date specified by datetype. Format: YYYY/MM/DD. Must be used with mindate.)'},\n",
       "   'functionality': 'Provides a list of UIDs matching a text query',\n",
       "   'env_requirements': ['requests'],\n",
       "   'example_code': \"import requests\\n\\n# Search PubMed for articles about cancer published in the last 60 days, retrieve the first 100 results, and return the results in JSON format\\nurl = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&term=cancer&reldate=60&datetype=edat&retmax=100&retmode=json'\\nresponse = requests.get(url)\\nprint(response.json())\",\n",
       "   'meta_data': {'description': 'The NCBI ESearch API provides a way to search Entrez databases using text queries and retrieve a list of UIDs matching the query.',\n",
       "    'documentation': 'https://www.ncbi.nlm.nih.gov/books/NBK25499/'},\n",
       "   'questions': ['How can I retrieve a list of PubMed IDs for articles published in the last year?',\n",
       "    'I need to find all the nucleotide sequences that have been modified in the last 30 days. How can I do this?',\n",
       "    \"I'm working on a research project about gene expression. I need to find all the genes that have been associated with a specific disease in the last 5 years. How can I use the ESearch API to retrieve this information?\",\n",
       "    \"I'm trying to find all the protein sequences that have been annotated with a specific function. How can I use the ESearch API to retrieve this information?\",\n",
       "    'I need to retrieve a list of all the articles published in a specific journal. How can I use the ESearch API to retrieve this information?']}],\n",
       " [{'user_name': 'template_usrname',\n",
       "   'api_name': 'EInfo API',\n",
       "   'api_call': 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/einfo.fcgi',\n",
       "   'api_version': '2.0',\n",
       "   'api_arguments': {'db': 'string (The name of the Entrez database to retrieve information about)',\n",
       "    'version': \"string (The version of the EInfo XML to return. The only supported value is '2.0')\",\n",
       "    'retmode': \"string (The retrieval type. Determines the format of the returned output. The default value is 'xml' for EInfo XML, but 'json' is also supported to return output in JSON format.)\"},\n",
       "   'functionality': 'Provides information about Entrez databases',\n",
       "   'env_requirements': ['requests'],\n",
       "   'example_code': \"import requests\\n\\n# Get a list of all Entrez database names\\nresponse = requests.get('https://eutils.ncbi.nlm.nih.gov/entrez/eutils/einfo.fcgi')\\nprint(response.text)\\n\\n# Get statistics for the Entrez Protein database in version 2.0 EInfo XML\\nresponse = requests.get('https://eutils.ncbi.nlm.nih.gov/entrez/eutils/einfo.fcgi?db=protein&version=2.0')\\nprint(response.text)\",\n",
       "   'meta_data': {'description': 'The EInfo API provides information about Entrez databases, including a list of all valid databases and statistics for a single database.',\n",
       "    'documentation': 'https://www.ncbi.nlm.nih.gov/books/NBK25499/'},\n",
       "   'questions': ['How can I obtain a list of all available Entrez databases?',\n",
       "    'What are the available link names for the PubMed database?',\n",
       "    'I need to retrieve information about the Entrez Gene database. What are the indexing fields available for this database?',\n",
       "    \"I'm working with the Entrez Protein database. How can I retrieve statistics for this database?\",\n",
       "    \"I need to use the wildcard character '*' in my search terms. Which Entrez databases support this feature?\",\n",
       "    'I need to search for entries within a specific date range. Which Entrez databases support this feature?']}],\n",
       " [{'user_name': 'template_usrname',\n",
       "   'api_name': 'ELink API',\n",
       "   'api_call': 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom={dbfrom}&db={db}&id={id}&cmd={cmd}&retmode={retmode}&linkname={linkname}&term={term}&holding={holding}&datetype={datetype}&reldate={reldate}&mindate={mindate}&maxdate={maxdate}',\n",
       "   'api_version': '1.0',\n",
       "   'api_arguments': {'dbfrom': 'string (The database containing the input UIDs. The value must be a valid Entrez database name (default = pubmed). This is the origin database of the link operation.)',\n",
       "    'db': 'string (Database from which to retrieve UIDs. The value must be a valid Entrez database name (default = pubmed). This is the destination database for the link operation.)',\n",
       "    'id': 'string (UID list. Either a single UID or a comma-delimited list of UIDs may be provided. All of the UIDs must be from the database specified by dbfrom. There is no set maximum for the number of UIDs that can be passed to ELink, but if more than about 200 UIDs are to be provided, the request should be made using the HTTP POST method.)',\n",
       "    'cmd': 'string (ELink command mode. The command mode specified which function ELink will perform. Some optional parameters only function for certain values of &cmd (see below).)',\n",
       "    'retmode': 'string (Retrieval type. Determines the format of the returned output. The default value is ‘xml’ for ELink XML, but ‘json’ is also supported to return output in JSON format.)',\n",
       "    'linkname': 'string (Name of the Entrez link to retrieve. Every link in Entrez is given a name of the form dbfrom_db_subset. The values of subset vary depending on the values of dbfrom and db. Many dbfrom/db combinations have no subset values. See the list of Entrez links for a listing of all available linknames. When linkname is used, only the links with that name will be retrieved.)',\n",
       "    'term': 'string (Entrez query used to limit the output set of linked UIDs. The query in the term parameter will be applied after the link operation, and only those UIDs matching the query will be returned by ELink. The term parameter only functions when db and dbfrom are set to the same database value.)',\n",
       "    'holding': \"string (Name of LinkOut provider. Only URLs for the LinkOut provider specified by holding will be returned. The value provided to holding should be the abbreviation of the LinkOut provider's name found in the <NameAbbr> tag of the ELink XML output when cmd is set to llinks or llinkslib. The holding parameter only functions when cmd is set to llinks or llinkslib.)\",\n",
       "    'datetype': \"string (Type of date used to limit a link operation. The allowed values vary between Entrez databases, but common values are 'mdat' (modification date), 'pdat' (publication date) and 'edat' (Entrez date). Generally an Entrez database will have only two allowed values for datetype.)\",\n",
       "    'reldate': 'integer (When reldate is set to an integer n, ELink returns only those items that have a date specified by datetype within the last n days.)',\n",
       "    'mindate': 'string (Date range used to limit a link operation by the date specified by datetype. These two parameters (mindate, maxdate) must be used together to specify an arbitrary date range. The general date format is YYYY/MM/DD, and these variants are also allowed: YYYY, YYYY/MM.)',\n",
       "    'maxdate': 'string (Date range used to limit a link operation by the date specified by datetype. These two parameters (mindate, maxdate) must be used together to specify an arbitrary date range. The general date format is YYYY/MM/DD, and these variants are also allowed: YYYY, YYYY/MM.)'},\n",
       "   'functionality': 'Returns UIDs linked to an input set of UIDs in either the same or a different Entrez database',\n",
       "   'env_requirements': ['requests'],\n",
       "   'example_code': \"import requests\\n\\n# Link from protein to gene\\napi_url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=protein&db=gene&id=15718680,157427902'\\nresponse = requests.get(api_url)\\nprint(response.text)\",\n",
       "   'meta_data': {'description': 'The ELink API allows users to retrieve UIDs linked to an input set of UIDs in either the same or a different Entrez database. It can also be used to return UIDs linked to other UIDs in the same Entrez database that match an Entrez query, check for the existence of Entrez links for a set of UIDs within the same database, list the available links for a UID, list LinkOut URLs and attributes for a set of UIDs, list hyperlinks to primary LinkOut providers for a set of UIDs, and create hyperlinks to the primary LinkOut provider for a single UID.',\n",
       "    'documentation': 'https://www.ncbi.nlm.nih.gov/books/NBK25499/'},\n",
       "   'questions': ['How can I find all the genes associated with a given protein?',\n",
       "    'I have a list of PubMed IDs. Can I find all the related articles that were published in the last year?',\n",
       "    'What are the available links for a specific PubMed ID?',\n",
       "    'I need to find all the articles that cite a specific paper. How can I do that?',\n",
       "    \"I'm working with a set of protein sequences. I need to find all the related sequences in the database. How can I do this?\"]}]]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "responses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
