{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import google.generativeai as genai\n",
    "import os\n",
    "import pandas as pd\n",
    "from IPython.display import display, Markdown, Latex\n",
    "from pathlib import Path\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "OUT_PATH = 'data/eutils_cleaned'\n",
    "DATA_DIR = \"data/eutils_raw/\"\n",
    "BASE_PROMPT_PATH = \"data/prompts/apidoc2md.md\"\n",
    "\n",
    "\n",
    "DOC_ORDER = [\"einfo\", \"esearch\", \"epost\",\n",
    "             \"esummary\", \"efetch\", \"elink\", \n",
    "             \"egquery\", \"espell\", \"ecitmatch\"]\n",
    "DOC_ORDER = pd.DataFrame(zip(DOC_ORDER, range(len(DOC_ORDER))), \n",
    "                         columns=[\"api_name\", \"format_order\"])\n",
    "\n",
    "CONTEXT_PATH = \"data/prompts/chap_4_tab_1.md\"\n",
    "\n",
    "Path(OUT_PATH).mkdir(parents=True, exist_ok=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "genai.configure(api_key=os.environ[\"GEMINI_API_KEY\"])\n",
    "\n",
    "# See https://ai.google.dev/api/python/google/generativeai/GenerativeModel\n",
    "# when response_mime_type is set to application/json api returns \"RECITATION\"\n",
    "# due to some repetition safe guard?\n",
    "# https://dropbox.tech/machine-learning/bye-bye-bye-evolution-of-repeated-token-attacks-on-chatgpt-models \n",
    "generation_config = {\n",
    "  \"temperature\": 0,\n",
    "  \"top_p\": 0.95,\n",
    "  \"top_k\": 64,\n",
    "  \"max_output_tokens\": 8192,\n",
    "  \"response_mime_type\": \"text/plain\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "model = genai.GenerativeModel(\n",
    "  model_name=\"gemini-1.5-flash-latest\",\n",
    "  generation_config=generation_config,\n",
    "  system_instruction=\"You are a helpful, thorough assistant. You analyze raw, unformatted text and output clean, [structured markdown](https://www.markdownguide.org/).\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "api_docs = []\n",
    "\n",
    "for _file in Path(DATA_DIR).iterdir():\n",
    "    with open(_file) as f:\n",
    "        doc = f.read()\n",
    "    api_docs.append([Path(_file).stem, doc])\n",
    "\n",
    "api_docs = pd.DataFrame(api_docs, columns=[\"api_name\", \"raw_text\"])\n",
    "api_docs = api_docs.merge(DOC_ORDER)\n",
    "api_docs = api_docs.sort_values([\"format_order\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(BASE_PROMPT_PATH) as f:\n",
    "    base_prompt = f.read()\n",
    "\n",
    "with open(CONTEXT_PATH) as f:\n",
    "    context = f.read()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 9/9 [01:54<00:00, 12.78s/it]\n"
     ]
    }
   ],
   "source": [
    "model = genai.GenerativeModel(\n",
    "  model_name=\"gemini-1.5-flash-latest\",\n",
    "  generation_config=generation_config,\n",
    ")\n",
    "\n",
    "\n",
    "responses = []\n",
    "current_docs = None\n",
    "for index, row in tqdm(api_docs.iterrows(), total=api_docs.shape[0]):\n",
    "  # tack on extra context if the document tells the readers\n",
    "  # to see tables. Avoid doing otherwise to avoid unnecessary \n",
    "  # inference costs\n",
    "  if row[\"api_name\"] == \"efetch\":\n",
    "    prompt = base_prompt.replace(\"{CONTEXT_RELEVANT}\", \"True\")\n",
    "    prompt = prompt.replace(\"{CONTEXT}\", context)\n",
    "  else:\n",
    "    prompt = base_prompt.replace(\"{CONTEXT_RELEVANT}\", \"False\")\n",
    "    prompt = prompt.replace(\"{CONTEXT}\", \"\")\n",
    "\n",
    "\n",
    "  response = model.generate_content([prompt, \n",
    "                                    f\"filename: {row['api_name']}.txt\", \n",
    "                                    row[\"raw_text\"]])\n",
    "  try:\n",
    "    responses.append(response.text)\n",
    "  except ValueError:\n",
    "    try: \n",
    "      print(response.candidates)\n",
    "    except Exception as e:\n",
    "      print(\"Gemini is throwing a tantrum...\")\n",
    "      print(f\"Exception {e} returned from API.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "api_docs[\"text_md\"] = responses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "## EInfo API Documentation\n",
       "\n",
       "### Base URL\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/einfo.fcgi\n",
       "```\n",
       "\n",
       "### Functions\n",
       "\n",
       "The EInfo API provides the following functionalities:\n",
       "\n",
       "* **List of all valid Entrez databases:** Returns a list of all valid Entrez database names when no `db` parameter is provided.\n",
       "* **Database statistics:** Provides statistics for a single database, including lists of indexing fields and available link names.\n",
       "\n",
       "### Required Parameters\n",
       "\n",
       "None. If no `db` parameter is provided, EInfo will return a list of the names of all valid Entrez databases.\n",
       "\n",
       "### Optional Parameters\n",
       "\n",
       "| Parameter | Description | Values |\n",
       "|---|---|---|\n",
       "| db | Target database about which to gather statistics. Value must be a valid Entrez database name. | annotinfo, assembly, biocollections, bioproject, biosample, blastdbinfo, books, cdd, clinvar, dbvar, gap, gapplus, gds, gene, genome, geoprofiles, grasp, gtr, ipg, medgen, mesh, nlmcatalog, nuccore, nucleotide, omim, orgtrack, pcassay, pccompound, pcsubstance, pmc, popset, protein, proteinclusters, protfam, pubmed, seqannot, snp, sra, structure, taxonomy |\n",
       "| version | Used to specify version 2.0 EInfo XML. The only supported value is '2.0'. When present, EInfo will return XML that includes two new fields: `<IsTruncatable>` and `<IsRangeable>`. Fields that are truncatable allow the wildcard character '*' in terms. The wildcard character will expand to match any set of characters up to a limit of 600 unique expansions. Fields that are rangeable allow the range operator ':' to be placed between a lower and upper limit for the desired range (e.g. 2008:2010[pdat]). | '2.0' |\n",
       "| retmode | Retrieval type. Determines the format of the returned output. The default value is 'xml' for EInfo XML, but 'json' is also supported to return output in JSON format. | 'xml', 'json' |\n",
       "\n",
       "### Examples\n",
       "\n",
       "**Return a list of all Entrez database names:**\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/einfo.fcgi\n",
       "```\n",
       "\n",
       "**Return version 2.0 statistics for Entrez Protein:**\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/einfo.fcgi?db=protein&version=2.0\n",
       "``` \n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "## ESearch API Documentation\n",
       "\n",
       "The ESearch API provides a way to search Entrez databases and retrieve a list of UIDs (Unique Identifiers) matching a text query. It also allows you to post search results to the History server, download UIDs from a dataset stored on the History server, combine or limit UID datasets, and sort sets of UIDs.\n",
       "\n",
       "**Base URL:** https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi\n",
       "\n",
       "**Note:** Some NCBI products contain search tools that generate content from searches on the web interface that are not available to ESearch. For example, the PubMed web interface (pubmed.ncbi.nlm.nih.gov) contains citation matching and spelling correction tools that are only available through that interface. Please see ECitMatch and ESpell for API equivalents.\n",
       "\n",
       "### Required Parameters\n",
       "\n",
       "| Parameter | Description | Values |\n",
       "|---|---|---|\n",
       "| db | Database to search. Value must be a valid Entrez database name. | annotinfo, assembly, biocollections, bioproject, biosample, blastdbinfo, books, cdd, clinvar, dbvar, gap, gapplus, gds, gene, genome, geoprofiles, grasp, gtr, ipg, medgen, mesh, nlmcatalog, nuccore, nucleotide, omim, orgtrack, pcassay, pccompound, pcsubstance, pmc, popset, protein, proteinclusters, protfam, pubmed, seqannot, snp, sra, structure, taxonomy |\n",
       "| term | Entrez text query. All special characters must be URL encoded. Spaces may be replaced by '+' signs. For very long queries (more than several hundred characters long), consider using an HTTP POST call. See the PubMed or Entrez help for information about search field descriptions and tags. Search fields and tags are database specific. | String |\n",
       "\n",
       "**Example:**\n",
       "\n",
       "```\n",
       "esearch.fcgi?db=pubmed&term=asthma\n",
       "```\n",
       "\n",
       "**Proximity Searching in PubMed:**\n",
       "\n",
       "PubMed also offers “proximity searching” for multiple terms appearing in any order within a specified number of words from one another in the [Title] or [Title/Abstract] fields.\n",
       "\n",
       "**Example:**\n",
       "\n",
       "```\n",
       "esearch.fcgi?db=pubmed&term=”asthma treatment”[Title:~3]\n",
       "```\n",
       "\n",
       "### Optional Parameters - History Server\n",
       "\n",
       "| Parameter | Description | Values |\n",
       "|---|---|---|\n",
       "| usehistory | When set to 'y', ESearch will post the UIDs resulting from the search operation onto the History server so that they can be used directly in a subsequent E-utility call. Also, usehistory must be set to 'y' for ESearch to interpret query key values included in term or to accept a WebEnv as input. | 'y' |\n",
       "| WebEnv | Web environment string returned from a previous ESearch, EPost or ELink call. When provided, ESearch will post the results of the search operation to this pre-existing WebEnv, thereby appending the results to the existing environment. In addition, providing WebEnv allows query keys to be used in term so that previous search sets can be combined or limited. As described above, if WebEnv is used, usehistory must be set to 'y'. | String |\n",
       "| query_key | Integer query key returned by a previous ESearch, EPost or ELink call. When provided, ESearch will find the intersection of the set specified by query_key and the set retrieved by the query in term (i.e. joins the two with AND). For query_key to function, WebEnv must be assigned an existing WebEnv string and usehistory must be set to 'y'. | Integer |\n",
       "\n",
       "**Example:**\n",
       "\n",
       "```\n",
       "esearch.fcgi?db=pubmed&term=asthma&WebEnv=<webenv string>&usehistory=y\n",
       "```\n",
       "\n",
       "**Query Keys in term:**\n",
       "\n",
       "Values for query keys may also be provided in term if they are preceeded by a '#' (%23 in the URL). While only one query_key parameter can be provided to ESearch, any number of query keys can be combined in term. Also, if query keys are provided in term, they can be combined with OR or NOT in addition to AND.\n",
       "\n",
       "**Example:**\n",
       "\n",
       "The following two URLs are functionally equivalent:\n",
       "\n",
       "```\n",
       "esearch.fcgi?db=pubmed&term=asthma&query_key=1&WebEnv=<webenv string>&usehistory=y\n",
       "\n",
       "esearch.fcgi?db=pubmed&term=%231+AND+asthma&WebEnv=<webenv string>&usehistory=y\n",
       "```\n",
       "\n",
       "### Optional Parameters - Retrieval\n",
       "\n",
       "| Parameter | Description | Values |\n",
       "|---|---|---|\n",
       "| retstart | Sequential index of the first UID in the retrieved set to be shown in the XML output (default=0, corresponding to the first record of the entire set). This parameter can be used in conjunction with retmax to download an arbitrary subset of UIDs retrieved from a search. | Integer |\n",
       "| retmax | Total number of UIDs from the retrieved set to be shown in the XML output (default=20). By default, ESearch only includes the first 20 UIDs retrieved in the XML output. If usehistory is set to 'y', the remainder of the retrieved set will be stored on the History server; otherwise these UIDs are lost. Increasing retmax allows more of the retrieved UIDs to be included in the XML output, up to a maximum of 10,000 records. | Integer |\n",
       "| rettype | Retrieval type. There are two allowed values for ESearch: 'uilist' (default), which displays the standard XML output, and 'count', which displays only the <Count> tag. | 'uilist', 'count' |\n",
       "| retmode | Retrieval type. Determines the format of the returned output. The default value is ‘xml’ for ESearch XML, but ‘json’ is also supported to return output in JSON format. | 'xml', 'json' |\n",
       "| sort | Specifies the method used to sort UIDs in the ESearch output. The available values vary by database (db) and may be found in the Display Settings menu on an Entrez search results page. If usehistory is set to ‘y’, the UIDs are loaded onto the History Server in the specified sort order and will be retrieved in that order by ESummary or EFetch. Example values are ‘relevance’ and ‘name’ for Gene. Users should be aware that the default value of sort varies from one database to another, and that the default value used by ESearch for a given database may differ from that used on NCBI web search pages. | Database specific |\n",
       "| field | Search field. If used, the entire search term will be limited to the specified Entrez field. | String |\n",
       "| idtype | Specifies the type of identifier to return for sequence databases (nuccore, popset, protein). By default, ESearch returns GI numbers in its output. If idtype is set to ‘acc’, ESearch will return accession.version identifiers rather than GI numbers. | 'acc', 'gi' |\n",
       "\n",
       "**Example:**\n",
       "\n",
       "```\n",
       "esearch.fcgi?db=pubmed&term=asthma&field=title\n",
       "```\n",
       "\n",
       "**Sort Values for PubMed:**\n",
       "\n",
       "* pub_date – descending sort by publication date\n",
       "* Author – ascending sort by first author\n",
       "* JournalName – ascending sort by journal name\n",
       "* relevance – default sort order, (“Best Match”) on web PubMed\n",
       "\n",
       "### Optional Parameters - Dates\n",
       "\n",
       "| Parameter | Description | Values |\n",
       "|---|---|---|\n",
       "| datetype | Type of date used to limit a search. The allowed values vary between Entrez databases, but common values are 'mdat' (modification date), 'pdat' (publication date) and 'edat' (Entrez date). Generally an Entrez database will have only two allowed values for datetype. | Database specific |\n",
       "| reldate | When reldate is set to an integer n, the search returns only those items that have a date specified by datetype within the last n days. | Integer |\n",
       "| mindate | Date range used to limit a search result by the date specified by datetype. These two parameters (mindate, maxdate) must be used together to specify an arbitrary date range. The general date format is YYYY/MM/DD, and these variants are also allowed: YYYY, YYYY/MM. | YYYY/MM/DD, YYYY, YYYY/MM |\n",
       "| maxdate | Date range used to limit a search result by the date specified by datetype. These two parameters (mindate, maxdate) must be used together to specify an arbitrary date range. The general date format is YYYY/MM/DD, and these variants are also allowed: YYYY, YYYY/MM. | YYYY/MM/DD, YYYY, YYYY/MM |\n",
       "\n",
       "**Example:**\n",
       "\n",
       "Search in PubMed with the term cancer for abstracts that have an Entrez date within the last 60 days; retrieve the first 100 PMIDs and translations; post the results on the History server and return a WebEnv and query_key:\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&term=cancer&reldate=60&datetype=edat&retmax=100&usehistory=y\n",
       "```\n",
       "\n",
       "### Examples\n",
       "\n",
       "* Search in PubMed for the journal PNAS, Volume 97, and retrieve six PMIDs starting with the seventh PMID in the list:\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&term=PNAS[ta]+AND+97[vi]&retstart=6&retmax=6&tool=biomed3\n",
       "```\n",
       "\n",
       "* Search in the NLM Catalog for journals matching the term obstetrics:\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=nlmcatalog&term=obstetrics+AND+ncbijournals[filter]\n",
       "```\n",
       "\n",
       "* Search PubMed Central for free full text articles containing the query stem cells:\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pmc&term=stem+cells+AND+free+fulltext[filter]\n",
       "```\n",
       "\n",
       "* Search in Nucleotide for all tRNAs:\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=nucleotide&term=biomol+trna[prop]\n",
       "```\n",
       "\n",
       "* Search in Protein for a molecular weight range:\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=protein&term=70000:90000[molecular+weight]\n",
       "```\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "## EPost API Documentation\n",
       "\n",
       "The EPost API allows you to upload a list of UIDs to the Entrez History server or append a list of UIDs to an existing set of UID lists attached to a Web Environment.\n",
       "\n",
       "### Base URL\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/epost.fcgi\n",
       "```\n",
       "\n",
       "### Functions\n",
       "\n",
       "* **Uploads a list of UIDs to the Entrez History server:** This function allows you to store a list of UIDs for later retrieval.\n",
       "* **Appends a list of UIDs to an existing set of UID lists attached to a Web Environment:** This function allows you to add a list of UIDs to an existing Web Environment, which can be used for further processing or analysis.\n",
       "\n",
       "### Required Parameters\n",
       "\n",
       "| Parameter | Description | Values |\n",
       "|---|---|---|\n",
       "| db | Database containing the UIDs in the input list. The value must be a valid Entrez database name. | annotinfo, assembly, biocollections, bioproject, biosample, blastdbinfo, books, cdd, clinvar, dbvar, gap, gapplus, gds, gene, genome, geoprofiles, grasp, gtr, ipg, medgen, mesh, nlmcatalog, nuccore, nucleotide, omim, orgtrack, pcassay, pccompound, pcsubstance, pmc, popset, protein, proteinclusters, protfam, pubmed, seqannot, snp, sra, structure, taxonomy |\n",
       "| id | UID list. Either a single UID or a comma-delimited list of UIDs may be provided. All of the UIDs must be from the database specified by db. For PubMed, no more than 10,000 UIDs can be included in a single URL request. For other databases there is no set maximum for the number of UIDs that can be passed to epost, but if more than about 200 UIDs are to be posted, the request should be made using the HTTP POST method. For sequence databases (nuccore, popset, protein), the UID list may be a mixed list of GI numbers and accession.version identifiers. Note: When using accession.version identifiers, there is a conversion step that takes place that causes large lists of identifiers to time out, even when using POST. Therefore, we recommend batching these types of requests in sizes of about 500 UIDs or less, to avoid retrieving only a partial amount of records from your original POST input list. | String (comma-separated list of UIDs) |\n",
       "\n",
       "### Optional Parameter\n",
       "\n",
       "| Parameter | Description | Values |\n",
       "|---|---|---|\n",
       "| WebEnv | Web Environment. If provided, this parameter specifies the Web Environment that will receive the UID list sent by post. EPost will create a new query key associated with that Web Environment. Usually this WebEnv value is obtained from the output of a previous ESearch, EPost or ELink call. If no WebEnv parameter is provided, EPost will create a new Web Environment and post the UID list to query_key 1. | String |\n",
       "\n",
       "### Example\n",
       "\n",
       "**Post records to PubMed:**\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/epost.fcgi?db=pubmed&id=11237011,12466850\n",
       "```\n",
       "\n",
       "**Post records to Protein database with a specific Web Environment:**\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/epost.fcgi?db=protein&id=15718680,157427902,119703751&WebEnv=<webenv string>\n",
       "```\n",
       "\n",
       "**Note:** Replace `<webenv string>` with the actual Web Environment value. \n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "## ESummary API Documentation\n",
       "\n",
       "### Base URL\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi\n",
       "```\n",
       "\n",
       "### Functions\n",
       "\n",
       "The ESummary API retrieves document summaries (DocSums) for a list of input UIDs or a set of UIDs stored on the Entrez History server.\n",
       "\n",
       "### Required Parameters\n",
       "\n",
       "#### Input from UID List\n",
       "\n",
       "| Parameter | Description | Values |\n",
       "|---|---|---|\n",
       "| db | Database from which to retrieve DocSums. | annotinfo, assembly, biocollections, bioproject, biosample, blastdbinfo, books, cdd, clinvar, dbvar, gap, gapplus, gds, gene, genome, geoprofiles, grasp, gtr, ipg, medgen, mesh, nlmcatalog, nuccore, nucleotide, omim, orgtrack, pcassay, pccompound, pcsubstance, pmc, popset, protein, proteinclusters, protfam, pubmed, seqannot, snp, sra, structure, taxonomy |\n",
       "| id | UID list. Either a single UID or a comma-delimited list of UIDs may be provided. All of the UIDs must be from the database specified by db. There is no set maximum for the number of UIDs that can be passed to ESummary, but if more than about 200 UIDs are to be provided, the request should be made using the HTTP POST method. For sequence databases (nuccore, popset, protein), the UID list may be a mixed list of GI numbers and accession.version identifiers. | String (comma-separated list of UIDs) |\n",
       "\n",
       "#### Input from Entrez History Server\n",
       "\n",
       "| Parameter | Description | Values |\n",
       "|---|---|---|\n",
       "| db | Database from which to retrieve DocSums. | annotinfo, assembly, biocollections, bioproject, biosample, blastdbinfo, books, cdd, clinvar, dbvar, gap, gapplus, gds, gene, genome, geoprofiles, grasp, gtr, ipg, medgen, mesh, nlmcatalog, nuccore, nucleotide, omim, orgtrack, pcassay, pccompound, pcsubstance, pmc, popset, protein, proteinclusters, protfam, pubmed, seqannot, snp, sra, structure, taxonomy |\n",
       "| query_key | Query key. This integer specifies which of the UID lists attached to the given Web Environment will be used as input to ESummary. Query keys are obtained from the output of previous ESearch, EPost or ELink calls. | Integer |\n",
       "| WebEnv | Web Environment. This parameter specifies the Web Environment that contains the UID list to be provided as input to ESummary. Usually this WebEnv value is obtained from the output of a previous ESearch, EPost or ELink call. | String |\n",
       "\n",
       "### Optional Parameters\n",
       "\n",
       "#### Retrieval\n",
       "\n",
       "| Parameter | Description | Values |\n",
       "|---|---|---|\n",
       "| retstart | Sequential index of the first DocSum to be retrieved (default=1, corresponding to the first record of the entire set). This parameter can be used in conjunction with retmax to download an arbitrary subset of DocSums from the input set. | Integer |\n",
       "| retmax | Total number of DocSums from the input set to be retrieved, up to a maximum of 10,000. If the total set is larger than this maximum, the value of retstart can be iterated while holding retmax constant, thereby downloading the entire set in batches of size retmax. | Integer |\n",
       "| retmode | Retrieval type. Determines the format of the returned output. The default value is ‘xml’ for ESummary XML, but ‘json’ is also supported to return output in JSON format. | xml, json |\n",
       "| version | Used to specify version 2.0 ESummary XML. The only supported value is ‘2.0’. When present, ESummary will return version 2.0 DocSum XML that is unique to each Entrez database and that often contains more data than the default DocSum XML. | 2.0 |\n",
       "\n",
       "### Examples\n",
       "\n",
       "#### PubMed\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=pubmed&id=11850928,11482001\n",
       "```\n",
       "\n",
       "#### PubMed, version 2.0 XML\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=pubmed&id=11850928,11482001&version=2.0\n",
       "```\n",
       "\n",
       "#### Protein\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=protein&id=28800982,28628843\n",
       "```\n",
       "\n",
       "#### Nucleotide\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=nucleotide&id=28864546,28800981\n",
       "```\n",
       "\n",
       "#### Structure\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=structure&id=19923,12120\n",
       "```\n",
       "\n",
       "#### Taxonomy\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=taxonomy&id=9913,30521\n",
       "```\n",
       "\n",
       "#### UniSTS\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=unists&id=254085,254086\n",
       "```\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "## EFetch API Documentation\n",
       "\n",
       "### Base URL\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi\n",
       "```\n",
       "\n",
       "### Functions\n",
       "\n",
       "The EFetch API retrieves formatted data records for a list of input UIDs or a set of UIDs stored on the Entrez History server.\n",
       "\n",
       "### Required Parameters\n",
       "\n",
       "#### Input from UID List\n",
       "\n",
       "| Parameter | Description | Values |\n",
       "|---|---|---|\n",
       "| db | Database from which to retrieve records. The value must be a valid Entrez database name (default = pubmed). | annotinfo, assembly, biocollections, bioproject, biosample, blastdbinfo, books, cdd, clinvar, dbvar, gap, gapplus, gds, gene, genome, geoprofiles, grasp, gtr, ipg, medgen, mesh, nlmcatalog, nuccore, nucleotide, omim, orgtrack, pcassay, pccompound, pcsubstance, pmc, popset, protein, proteinclusters, protfam, pubmed, seqannot, snp, sra, structure, taxonomy |\n",
       "| id | UID list. Either a single UID or a comma-delimited list of UIDs may be provided. All of the UIDs must be from the database specified by db. There is no set maximum for the number of UIDs that can be passed to EFetch, but if more than about 200 UIDs are to be provided, the request should be made using the HTTP POST method. For sequence databases (nuccore, popset, protein), the UID list may be a mixed list of GI numbers and accession.version identifiers. | String (comma-separated list of UIDs) |\n",
       "\n",
       "**Example:**\n",
       "\n",
       "```\n",
       "efetch.fcgi?db=pubmed&id=19393038,30242208,29453458\n",
       "efetch.fcgi?db=protein&id=15718680,NP_001098858.1,119703751\n",
       "```\n",
       "\n",
       "**Special Note for Sequence Databases:**\n",
       "\n",
       "NCBI is no longer assigning GI numbers to a growing number of new sequence records. These records are not indexed in Entrez and cannot be retrieved using ESearch or ESummary. They also have no Entrez links accessible by ELink. EFetch can retrieve these records by including their accession.version identifier in the `id` parameter.\n",
       "\n",
       "#### Input from Entrez History Server\n",
       "\n",
       "| Parameter | Description | Values |\n",
       "|---|---|---|\n",
       "| db | Database from which to retrieve records. The value must be a valid Entrez database name (default = pubmed). | annotinfo, assembly, biocollections, bioproject, biosample, blastdbinfo, books, cdd, clinvar, dbvar, gap, gapplus, gds, gene, genome, geoprofiles, grasp, gtr, ipg, medgen, mesh, nlmcatalog, nuccore, nucleotide, omim, orgtrack, pcassay, pccompound, pcsubstance, pmc, popset, protein, proteinclusters, protfam, pubmed, seqannot, snp, sra, structure, taxonomy |\n",
       "| query_key | Query key. This integer specifies which of the UID lists attached to the given Web Environment will be used as input to EFetch. Query keys are obtained from the output of previous ESearch, EPost or ELInk calls. The `query_key` parameter must be used in conjunction with `WebEnv`. | Integer |\n",
       "| WebEnv | Web Environment. This parameter specifies the Web Environment that contains the UID list to be provided as input to EFetch. Usually this `WebEnv` value is obtained from the output of a previous ESearch, EPost or ELink call. The `WebEnv` parameter must be used in conjunction with `query_key`. | String |\n",
       "\n",
       "**Example:**\n",
       "\n",
       "```\n",
       "efetch.fcgi?db=protein&query_key=<key>&WebEnv=<webenv string>\n",
       "```\n",
       "\n",
       "### Optional Parameters\n",
       "\n",
       "#### Retrieval\n",
       "\n",
       "| Parameter | Description | Values |\n",
       "|---|---|---|\n",
       "| retmode | Retrieval mode. This parameter specifies the data format of the records returned, such as plain text, HTML or XML. See Table 1 for a full list of allowed values for each database. | xml, text, asn.1, default |\n",
       "| rettype | Retrieval type. This parameter specifies the record view returned, such as Abstract or MEDLINE from PubMed, or GenPept or FASTA from protein. Please see Table 1 for a full list of allowed values for each database. | docsum, uilist, full, summary, gene_table, alignmentscores, fasta, homologene, null, native, acc, seqid, gb, gbc, ft, gbwithparts, fasta_cds_na, fasta_cds_aa, gp, gpc, ipg, medline, abstract, flt, rsr, ssexemplar, chr, docset, clinvarset, gtracc |\n",
       "| retstart | Sequential index of the first record to be retrieved (default=0, corresponding to the first record of the entire set). This parameter can be used in conjunction with `retmax` to download an arbitrary subset of records from the input set. | Integer |\n",
       "| retmax | Total number of records from the input set to be retrieved, up to a maximum of 10,000. Optionally, for a large set the value of `retstart` can be iterated while holding `retmax` constant, thereby downloading the entire set in batches of size `retmax`. | Integer |\n",
       "\n",
       "#### Sequence Databases\n",
       "\n",
       "| Parameter | Description | Values |\n",
       "|---|---|---|\n",
       "| strand | Strand of DNA to retrieve. Available values are \"1\" for the plus strand and \"2\" for the minus strand. | 1, 2 |\n",
       "| seq_start | First sequence base to retrieve. The value should be the integer coordinate of the first desired base, with \"1\" representing the first base of the sequence. | Integer |\n",
       "| seq_stop | Last sequence base to retrieve. The value should be the integer coordinate of the last desired base, with \"1\" representing the first base of the sequence. | Integer |\n",
       "| complexity | Integer value 0 through 4. Data content to return. Many sequence records are part of a larger data structure or \"blob\", and the `complexity` parameter determines how much of that blob to return. For example, an mRNA may be stored together with its protein product. | 0, 1, 2, 3, 4 |\n",
       "\n",
       "**Value of complexity data returned for each requested GI:**\n",
       "\n",
       "* 0 - entire blob\n",
       "* 1 - bioseq\n",
       "* 2 - minimal bioseq-set\n",
       "* 3 - minimal nuc-prot\n",
       "* 4 - minimal pub-set\n",
       "\n",
       "### Examples\n",
       "\n",
       "#### PubMed\n",
       "\n",
       "* Fetch PMIDs 17284678 and 9997 as text abstracts:\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=pubmed&id=17284678,9997&retmode=text&rettype=abstract\n",
       "```\n",
       "\n",
       "* Fetch PMIDs in XML:\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=pubmed&id=11748933,11700088&retmode=xml\n",
       "```\n",
       "\n",
       "#### PubMed Central\n",
       "\n",
       "* Fetch XML for PubMed Central ID 212403:\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=pmc&id=212403\n",
       "```\n",
       "\n",
       "#### Nucleotide/Nuccore\n",
       "\n",
       "* Fetch the first 100 bases of the plus strand of GI 21614549 in FASTA format:\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nuccore&id=21614549&strand=1&seq_start=1&seq_stop=100&rettype=fasta&retmode=text\n",
       "```\n",
       "\n",
       "* Fetch the first 100 bases of the minus strand of GI 21614549 in FASTA format:\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nuccore&id=21614549&strand=2&seq_start=1&seq_stop=100&rettype=fasta&retmode=text\n",
       "```\n",
       "\n",
       "* Fetch the nuc-prot object for GI 21614549:\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nuccore&id=21614549&complexity=3\n",
       "```\n",
       "\n",
       "* Fetch the full ASN.1 record for GI 5:\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&id=5\n",
       "```\n",
       "\n",
       "* Fetch FASTA for GI 5:\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&id=5&rettype=fasta\n",
       "```\n",
       "\n",
       "* Fetch the GenBank flat file for GI 5:\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&id=5&rettype=gb\n",
       "```\n",
       "\n",
       "* Fetch GBSeqXML for GI 5:\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&id=5&rettype=gb&retmode=xml\n",
       "```\n",
       "\n",
       "* Fetch TinySeqXML for GI 5:\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&id=5&rettype=fasta&retmode=xml\n",
       "```\n",
       "\n",
       "#### Popset\n",
       "\n",
       "* Fetch the GenPept flat file for Popset ID 12829836:\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=popset&id=12829836&rettype=gp\n",
       "```\n",
       "\n",
       "#### Protein\n",
       "\n",
       "* Fetch the GenPept flat file for GI 8:\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=protein&id=8&rettype=gp\n",
       "```\n",
       "\n",
       "* Fetch GBSeqXML for GI 8:\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=protein&id=8&rettype=gp&retmode=xml\n",
       "```\n",
       "\n",
       "#### Sequences\n",
       "\n",
       "* Fetch FASTA for a transcript and its protein product (GIs 312836839 and 34577063)\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=sequences&id=312836839,34577063&rettype=fasta&retmode=text\n",
       "```\n",
       "\n",
       "#### Gene\n",
       "\n",
       "* Fetch full XML record for Gene ID 2:\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=gene&id=2&retmode=xml\n",
       "```\n",
       "\n",
       "### Table 1: Valid values of `&retmode` and `&rettype` for EFetch (null = empty string)\n",
       "\n",
       "**All Databases**\n",
       "\n",
       "| Record Type | `&rettype` | `&retmode` |\n",
       "|---|---|---|\n",
       "| Document summary | docsum | xml, default |\n",
       "| List of UIDs in XML | uilist | xml |\n",
       "| List of UIDs in plain text | uilist | text |\n",
       "\n",
       "**db = bioproject**\n",
       "\n",
       "| Record Type | `&rettype` | `&retmode` |\n",
       "|---|---|---|\n",
       "| Full record XML | xml, default | xml, default |\n",
       "\n",
       "**db = biosample**\n",
       "\n",
       "| Record Type | `&rettype` | `&retmode` |\n",
       "|---|---|---|\n",
       "| Full record XML | full, default | xml, default |\n",
       "| Full record text | full, default | text |\n",
       "\n",
       "**db = gds**\n",
       "\n",
       "| Record Type | `&rettype` | `&retmode` |\n",
       "|---|---|---|\n",
       "| Summary | summary, default | text, default |\n",
       "\n",
       "**db = gene**\n",
       "\n",
       "| Record Type | `&rettype` | `&retmode` |\n",
       "|---|---|---|\n",
       "| text ASN.1 | null | asn.1, default |\n",
       "| XML | null | xml |\n",
       "| Gene table | gene_table | text |\n",
       "\n",
       "**db = homologene**\n",
       "\n",
       "| Record Type | `&rettype` | `&retmode` |\n",
       "|---|---|---|\n",
       "| text ASN.1 | null | asn.1, default |\n",
       "| XML | null | xml |\n",
       "| Alignment scores | alignmentscores | text |\n",
       "| FASTA | fasta | text |\n",
       "| HomoloGene | homologene | text |\n",
       "\n",
       "**db = mesh**\n",
       "\n",
       "| Record Type | `&rettype` | `&retmode` |\n",
       "|---|---|---|\n",
       "| Full record | full, default | text, default |\n",
       "\n",
       "**db = nlmcatalog**\n",
       "\n",
       "| Record Type | `&rettype` | `&retmode` |\n",
       "|---|---|---|\n",
       "| Full record | null | text, default |\n",
       "| XML | null | xml |\n",
       "\n",
       "**db = nuccore, protein or popset**\n",
       "\n",
       "| Record Type | `&rettype` | `&retmode` |\n",
       "|---|---|---|\n",
       "| text ASN.1 | null | text, default |\n",
       "| binary ASN.1 | null | asn.1 |\n",
       "| Full record in XML | native | xml |\n",
       "| Accession number(s) | acc | text |\n",
       "| FASTA | fasta | text |\n",
       "| TinySeq XML | fasta | xml |\n",
       "| SeqID string | seqid | text |\n",
       "\n",
       "**Additional options for db = nuccore or popset**\n",
       "\n",
       "| Record Type | `&rettype` | `&retmode` |\n",
       "|---|---|---|\n",
       "| GenBank flat file | gb | text |\n",
       "| GBSeq XML | gb | xml |\n",
       "| INSDSeq XML | gbc | xml |\n",
       "\n",
       "**Additional option for db = nuccore and protein**\n",
       "\n",
       "| Record Type | `&rettype` | `&retmode` |\n",
       "|---|---|---|\n",
       "| Feature table | ft | text |\n",
       "\n",
       "**Additional option for db = nuccore**\n",
       "\n",
       "| Record Type | `&rettype` | `&retmode` |\n",
       "|---|---|---|\n",
       "| GenBank flat file with full sequence (contigs) | gbwithparts | text |\n",
       "| CDS nucleotide FASTA | fasta_cds_na | text |\n",
       "| CDS protein FASTA | fasta_cds_aa | text |\n",
       "\n",
       "**Additional options for db = protein**\n",
       "\n",
       "| Record Type | `&rettype` | `&retmode` |\n",
       "|---|---|---|\n",
       "| GenPept flat file | gp | text |\n",
       "| GBSeq XML | gp | xml |\n",
       "| INSDSeq XML | gpc | xml |\n",
       "| Identical Protein XML | ipg | xml |\n",
       "\n",
       "**db = pmc**\n",
       "\n",
       "| Record Type | `&rettype` | `&retmode` |\n",
       "|---|---|---|\n",
       "| XML | null | xml, default |\n",
       "| MEDLINE | medline | text |\n",
       "\n",
       "**db = pubmed**\n",
       "\n",
       "| Record Type | `&rettype` | `&retmode` |\n",
       "|---|---|---|\n",
       "| XML | null | xml, default |\n",
       "| MEDLINE | medline | text |\n",
       "| PMID list | uilist | text |\n",
       "| Abstract | abstract | text |\n",
       "\n",
       "**db = sequences**\n",
       "\n",
       "| Record Type | `&rettype` | `&retmode` |\n",
       "|---|---|---|\n",
       "| text ASN.1 | null | text, default |\n",
       "| Accession number(s) | acc | text |\n",
       "| FASTA | fasta | text |\n",
       "| SeqID string | seqid | text |\n",
       "\n",
       "**db = snp**\n",
       "\n",
       "| Record Type | `&rettype` | `&retmode` |\n",
       "|---|---|---|\n",
       "| text ASN.1 | null | asn.1, default |\n",
       "| XML | null | xml |\n",
       "| Flat file | flt | text |\n",
       "| FASTA | fasta | text |\n",
       "| RS Cluster report | rsr | text |\n",
       "| SS Exemplar list | ssexemplar | text |\n",
       "| Chromosome report | chr | text |\n",
       "| Summary | docset | text |\n",
       "| UID list | uilist | text or xml |\n",
       "\n",
       "**db = sra**\n",
       "\n",
       "| Record Type | `&rettype` | `&retmode` |\n",
       "|---|---|---|\n",
       "| XML | full, default | xml, default |\n",
       "\n",
       "**db = taxonomy**\n",
       "\n",
       "| Record Type | `&rettype` | `&retmode` |\n",
       "|---|---|---|\n",
       "| XML | null | xml, default |\n",
       "| TaxID list | uilist | text or xml |\n",
       "\n",
       "**db = clinvar**\n",
       "\n",
       "| Record Type | `&rettype` | `&retmode` |\n",
       "|---|---|---|\n",
       "| ClinVar Set | clinvarset | xml, default |\n",
       "| UID list | uilist | text or xml |\n",
       "\n",
       "**db = gtr**\n",
       "\n",
       "| Record Type | `&rettype` | `&retmode` |\n",
       "|---|---|---|\n",
       "| GTR Test Report | gtracc | xml, default |\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "## ELink API Documentation\n",
       "\n",
       "### Base URL\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi\n",
       "```\n",
       "\n",
       "### Functions\n",
       "\n",
       "The ELink API provides functions for retrieving linked UIDs from different Entrez databases, finding related UIDs within the same database, checking for the existence of links, and listing LinkOut URLs.\n",
       "\n",
       "### Required Parameters\n",
       "\n",
       "| Parameter | Description | Values |\n",
       "|---|---|---|\n",
       "| db | Target database for the link operation. | annotinfo, assembly, biocollections, bioproject, biosample, blastdbinfo, books, cdd, clinvar, dbvar, gap, gapplus, gds, gene, genome, geoprofiles, grasp, gtr, ipg, medgen, mesh, nlmcatalog, nuccore, nucleotide, omim, orgtrack, pcassay, pccompound, pcsubstance, pmc, popset, protein, proteinclusters, protfam, pubmed, seqannot, snp, sra, structure, taxonomy |\n",
       "| dbfrom | Origin database containing the input UIDs. | annotinfo, assembly, biocollections, bioproject, biosample, blastdbinfo, books, cdd, clinvar, dbvar, gap, gapplus, gds, gene, genome, geoprofiles, grasp, gtr, ipg, medgen, mesh, nlmcatalog, nuccore, nucleotide, omim, orgtrack, pcassay, pccompound, pcsubstance, pmc, popset, protein, proteinclusters, protfam, pubmed, seqannot, snp, sra, structure, taxonomy |\n",
       "| cmd | ELink command mode. Specifies the function to be performed. | neighbor, neighbor_score, neighbor_history, acheck, ncheck, lcheck, llinks, llinkslib, prlinks |\n",
       "| id | UID list (single UID or comma-delimited list). | Integer |\n",
       "| query_key | Query key from a previous ESearch, EPost, or ELInk call. | Integer |\n",
       "| WebEnv | Web Environment from a previous ESearch, EPost, or ELInk call. | String |\n",
       "\n",
       "### Optional Parameters\n",
       "\n",
       "#### Retrieval\n",
       "\n",
       "| Parameter | Description | Values |\n",
       "|---|---|---|\n",
       "| retmode | Retrieval type. | xml, json |\n",
       "| idtype | Identifier type for sequence databases (nuccore, popset, protein). | gi, acc |\n",
       "\n",
       "#### Limiting the Output Set of Links\n",
       "\n",
       "| Parameter | Description | Values |\n",
       "|---|---|---|\n",
       "| linkname | Name of the Entrez link to retrieve. | String (e.g., gene_snp_genegenotype) |\n",
       "| term | Entrez query to limit the output set of linked UIDs. | String |\n",
       "| holding | Name of LinkOut provider. | String (e.g., CTgov) |\n",
       "\n",
       "#### Dates (Only for `cmd=neighbor` or `cmd=neighbor_history` and `dbfrom=pubmed`)\n",
       "\n",
       "| Parameter | Description | Values |\n",
       "|---|---|---|\n",
       "| datetype | Type of date used to limit the link operation. | mdat, pdat, edat |\n",
       "| reldate | Number of days to limit the link operation by the date specified by `datetype`. | Integer |\n",
       "| mindate | Minimum date for the date range. | YYYY/MM/DD, YYYY, YYYY/MM |\n",
       "| maxdate | Maximum date for the date range. | YYYY/MM/DD, YYYY, YYYY/MM |\n",
       "\n",
       "### Command Modes\n",
       "\n",
       "#### `cmd=neighbor` (Default)\n",
       "\n",
       "Returns a set of UIDs in `db` linked to the input UIDs in `dbfrom`.\n",
       "\n",
       "**Example:** Link from protein to gene\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=protein&db=gene&id=15718680,157427902\n",
       "```\n",
       "\n",
       "#### `cmd=neighbor_score`\n",
       "\n",
       "Returns a set of UIDs within the same database as the input UIDs along with computed similarity scores.\n",
       "\n",
       "**Example:** Find related articles to PMID 20210808\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=pubmed&db=pubmed&id=20210808&cmd=neighbor_score\n",
       "```\n",
       "\n",
       "#### `cmd=neighbor_history`\n",
       "\n",
       "Posts the output UIDs to the Entrez History server and returns a `query_key` and `WebEnv` corresponding to the location of the output set.\n",
       "\n",
       "**Example:** Link from protein to gene and post the results on the Entrez History\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=protein&db=gene&id=15718680,157427902&cmd=neighbor_history\n",
       "```\n",
       "\n",
       "#### `cmd=acheck`\n",
       "\n",
       "Lists all links available for a set of UIDs.\n",
       "\n",
       "**Example:** List all possible links from two protein GIs\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=protein&id=15718680,157427902&cmd=acheck\n",
       "```\n",
       "\n",
       "#### `cmd=ncheck`\n",
       "\n",
       "Checks for the existence of links within the same database for a set of UIDs.\n",
       "\n",
       "**Example:** Check whether two nuccore sequences have \"related sequences\" links.\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=nuccore&id=21614549,219152114&cmd=ncheck\n",
       "```\n",
       "\n",
       "#### `cmd=lcheck`\n",
       "\n",
       "Checks for the existence of external links (LinkOuts) for a set of UIDs.\n",
       "\n",
       "**Example:** Check whether two protein sequences have any LinkOut providers.\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=protein&id=15718680,157427902&cmd=lcheck\n",
       "```\n",
       "\n",
       "#### `cmd=llinks`\n",
       "\n",
       "Lists the URLs and attributes for the LinkOut providers that are not libraries for each input UID.\n",
       "\n",
       "**Example:** List the LinkOut URLs for non-library providers for two pubmed abstracts.\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=pubmed&id=19880848,19822630&cmd=llinks\n",
       "```\n",
       "\n",
       "#### `cmd=llinkslib`\n",
       "\n",
       "Lists the URLs and attributes for all LinkOut providers, including libraries, for each input UID.\n",
       "\n",
       "**Example:** List all LinkOut URLs for two PubMed abstracts.\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=pubmed&id=19880848,19822630&cmd=llinkslib\n",
       "```\n",
       "\n",
       "#### `cmd=prlinks`\n",
       "\n",
       "Lists the primary LinkOut provider for each input UID, or links directly to the LinkOut provider's web site for a single UID if `retmode` is set to `ref`.\n",
       "\n",
       "**Example:** Find links to full text providers for two PubMed abstracts.\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=pubmed&id=19880848,19822630&cmd=prlinks\n",
       "```\n",
       "\n",
       "**Example:** Link directly to the full text for a PubMed abstract at the provider's web site.\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=pubmed&id=19880848&cmd=prlinks&retmode=ref\n",
       "```\n",
       "\n",
       "### Notes\n",
       "\n",
       "* If more than one `id` parameter is provided, ELink will perform a separate link operation for each set of UIDs specified by each `id` parameter.\n",
       "* For sequence databases (nuccore, popset, protein), the `id` list may be a mixed list of GI numbers and accession.version identifiers.\n",
       "* The `linkname` parameter only functions when `cmd` is set to `neighbor` or `neighbor_history`.\n",
       "* The `term` parameter only functions when `db` and `dbfrom` are set to the same database value.\n",
       "* The `holding` parameter only functions when `cmd` is set to `llinks` or `llinkslib`.\n",
       "* The date parameters (`datetype`, `reldate`, `mindate`, `maxdate`) only function when `cmd` is set to `neighbor` or `neighbor_history` and `dbfrom` is `pubmed`.\n",
       "\n",
       "### Example Usage\n",
       "\n",
       "**Find all links from gene to snp:**\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=gene&db=snp&id=93986\n",
       "```\n",
       "\n",
       "**Find snps with genotype data linked to genes:**\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=gene&db=snp&id=93986&linkname=gene_snp_genegenotype\n",
       "```\n",
       "\n",
       "**Find all related articles for a PMID:**\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=pubmed&db=pubmed&id=19879512\n",
       "```\n",
       "\n",
       "**Find all related review articles published in 2008 for a PMID:**\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=pubmed&db=pubmed&id=19879512&term=review%5Bfilter%5D+AND+2008%5Bpdat%5Dh\n",
       "```\n",
       "\n",
       "**Find information for all LinkOut providers for a PMID:**\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=pubmed&cmd=llinkslib&id=16210666\n",
       "```\n",
       "\n",
       "**Find information from clinicaltrials.gov for a PMID:**\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=pubmed&cmd=llinkslib&id=16210666&holding=CTgov\n",
       "```\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "## EGQuery API Documentation\n",
       "\n",
       "### Base URL\n",
       "\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/egquery.fcgi\n",
       "\n",
       "### Function\n",
       "\n",
       "Provides the number of records retrieved in all Entrez databases by a single text query.\n",
       "\n",
       "### Required Parameters\n",
       "\n",
       "| Parameter | Description | Values |\n",
       "|---|---|---|\n",
       "| term | Entrez text query. All special characters must be URL encoded. Spaces may be replaced by '+' signs. For very long queries (more than several hundred characters long), consider using an HTTP POST call. See the PubMed or Entrez help for information about search field descriptions and tags. Search fields and tags are database specific. | String |\n",
       "\n",
       "### Example\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/egquery.fcgi?term=asthma\n",
       "``` \n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "## ESpell API Documentation\n",
       "\n",
       "**Base URL:** https://eutils.ncbi.nlm.nih.gov/entrez/eutils/espell.fcgi\n",
       "\n",
       "**Function:** Provides spelling suggestions for terms within a single text query in a given database.\n",
       "\n",
       "### Required Parameters\n",
       "\n",
       "| Parameter | Description | Values |\n",
       "|---|---|---|\n",
       "| db | Database to search. Value must be a valid Entrez database name. | annotinfo, assembly, biocollections, bioproject, biosample, blastdbinfo, books, cdd, clinvar, dbvar, gap, gapplus, gds, gene, genome, geoprofiles, grasp, gtr, ipg, medgen, mesh, nlmcatalog, nuccore, nucleotide, omim, orgtrack, pcassay, pccompound, pcsubstance, pmc, popset, protein, proteinclusters, protfam, pubmed, seqannot, snp, sra, structure, taxonomy |\n",
       "| term | Entrez text query. All special characters must be URL encoded. Spaces may be replaced by '+' signs. For very long queries (more than several hundred characters long), consider using an HTTP POST call. See the PubMed or Entrez help for information about search field descriptions and tags. Search fields and tags are database specific. | String |\n",
       "\n",
       "**Example:**\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/espell.fcgi?db=pubmed&term=asthmaa+OR+alergies\n",
       "``` \n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "## ECitMatch\n",
       "\n",
       "**Base URL:** https://eutils.ncbi.nlm.nih.gov/entrez/eutils/ecitmatch.cgi\n",
       "\n",
       "**Function:** Retrieves PubMed IDs (PMIDs) that correspond to a set of input citation strings.\n",
       "\n",
       "### Required Parameters\n",
       "\n",
       "| Parameter | Description | Values |\n",
       "|---|---|---|\n",
       "| db | Database to search. | pubmed |\n",
       "| rettype | Retrieval type. | xml |\n",
       "| bdata | Citation strings. Each input citation must be represented by a citation string in the following format:  `journal_title|year|volume|first_page|author_name|your_key|`  Multiple citation strings may be provided by separating the strings with a carriage return character (`%0D`). The `your_key` value is an arbitrary label provided by the user that may serve as a local identifier for the citation, and it will be included in the output. Be aware that all spaces must be replaced by '+' symbols and that citation strings should end with a final vertical bar '|'. | String |\n",
       "\n",
       "**Example:**\n",
       "\n",
       "```\n",
       "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/ecitmatch.cgi?db=pubmed&retmode=xml&bdata=proc+natl+acad+sci+u+s+a|1991|88|3248|mann+bj|Art1|%0Dscience|1987|235|182|palmenberg+ac|Art2|\n",
       "``` \n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for i in api_docs[\"text_md\"]:\n",
    "    display(Markdown(i))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "api_docs.to_parquet(\"data/entrez/markdown/chapter_4.parquet\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
