{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pydoc\n",
    "from Bio import Entrez\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python Library Documentation: package Bio.Entrez in Bio\n",
      "\n",
      "NAME\n",
      "    Bio.Entrez - Provides code to access NCBI over the WWW.\n",
      "\n",
      "DESCRIPTION\n",
      "    The main Entrez web page is available at:\n",
      "    http://www.ncbi.nlm.nih.gov/Entrez/\n",
      "\n",
      "    Entrez Programming Utilities web page is available at:\n",
      "    http://www.ncbi.nlm.nih.gov/books/NBK25501/\n",
      "\n",
      "    This module provides a number of functions like ``efetch`` (short for\n",
      "    Entrez Fetch) which will return the data as a handle object. This is\n",
      "    a standard interface used in Python for reading data from a file, or\n",
      "    in this case a remote network connection, and provides methods like\n",
      "    ``.read()`` or offers iteration over the contents line by line. See\n",
      "    also \"What the heck is a handle?\" in the Biopython Tutorial and\n",
      "    Cookbook: http://biopython.org/DIST/docs/tutorial/Tutorial.html\n",
      "    http://biopython.org/DIST/docs/tutorial/Tutorial.pdf\n",
      "    The handle returned by these functions can be either in text mode or\n",
      "    in binary mode, depending on the data requested and the results\n",
      "    returned by NCBI Entrez. Typically, XML data will be in binary mode\n",
      "    while other data will be in text mode, as required by the downstream\n",
      "    parser to parse the data.\n",
      "\n",
      "    Unlike a handle to a file on disk from the ``open(filename)`` function,\n",
      "    which has a ``.name`` attribute giving the filename, the handles from\n",
      "    ``Bio.Entrez`` all have a ``.url`` attribute instead giving the URL\n",
      "    used to connect to the NCBI Entrez API.\n",
      "\n",
      "    The ``epost``, ``efetch``, and ``esummary`` tools take an \"id\" parameter\n",
      "    which corresponds to one or more database UIDs (or accession.version\n",
      "    identifiers in the case of sequence databases such as \"nuccore\" or\n",
      "    \"protein\"). The Python value of the \"id\" keyword passed to these functions\n",
      "    may be either a single ID as a string or integer or multiple IDs as an\n",
      "    iterable of strings/integers. You may also pass a single string containing\n",
      "    multiple IDs delimited by commas. The ``elink`` tool also accepts multiple\n",
      "    IDs but the argument is handled differently than the other three. See that\n",
      "    function's docstring for more information.\n",
      "\n",
      "    All the functions that send requests to the NCBI Entrez API will\n",
      "    automatically respect the NCBI rate limit (of 3 requests per second\n",
      "    without an API key, or 10 requests per second with an API key) and\n",
      "    will automatically retry when encountering transient failures\n",
      "    (i.e. connection failures or HTTP 5XX codes). By default, Biopython\n",
      "    does a maximum of three tries before giving up, and sleeps for 15\n",
      "    seconds between tries. You can tweak these parameters by setting\n",
      "    ``Bio.Entrez.max_tries`` and ``Bio.Entrez.sleep_between_tries``.\n",
      "\n",
      "    The Entrez module also provides an XML parser which takes a handle\n",
      "    as input.\n",
      "\n",
      "    Variables:\n",
      "\n",
      "        - email        Set the Entrez email parameter (default is not set).\n",
      "        - tool         Set the Entrez tool parameter (default is ``biopython``).\n",
      "        - api_key      Personal API key from NCBI. If not set, only 3 queries per\n",
      "          second are allowed. 10 queries per seconds otherwise with a\n",
      "          valid API key.\n",
      "        - max_tries    Configures how many times failed requests will be\n",
      "          automatically retried on error (default is 3).\n",
      "        - sleep_between_tries   The delay, in seconds, before retrying a request on\n",
      "          error (default is 15).\n",
      "\n",
      "    Functions:\n",
      "\n",
      "        - efetch       Retrieves records in the requested format from a list of one or\n",
      "          more primary IDs or from the user's environment\n",
      "        - epost        Posts a file containing a list of primary IDs for future use in\n",
      "          the user's environment to use with subsequent search strategies\n",
      "        - esearch      Searches and retrieves primary IDs (for use in EFetch, ELink,\n",
      "          and ESummary) and term translations and optionally retains\n",
      "          results for future use in the user's environment.\n",
      "        - elink        Checks for the existence of an external or Related Articles link\n",
      "          from a list of one or more primary IDs.  Retrieves primary IDs\n",
      "          and relevancy scores for links to Entrez databases or Related\n",
      "          Articles;  creates a hyperlink to the primary LinkOut provider\n",
      "          for a specific ID and database, or lists LinkOut URLs\n",
      "          and Attributes for multiple IDs.\n",
      "        - einfo        Provides field index term counts, last update, and available\n",
      "          links for each database.\n",
      "        - esummary     Retrieves document summaries from a list of primary IDs or from\n",
      "          the user's environment.\n",
      "        - egquery      Provides Entrez database counts in XML for a single search\n",
      "          using Global Query.\n",
      "        - espell       Retrieves spelling suggestions.\n",
      "        - ecitmatch    Retrieves PubMed IDs (PMIDs) that correspond to a set of\n",
      "          input citation strings.\n",
      "\n",
      "        - read         Parses the XML results returned by any of the above functions.\n",
      "          Alternatively, the XML data can be read from a file opened in binary mode.\n",
      "          Typical usage is:\n",
      "\n",
      "              >>> from Bio import Entrez\n",
      "              >>> Entrez.email = \"Your.Name.Here@example.org\"\n",
      "              >>> handle = Entrez.einfo() # or esearch, efetch, ...\n",
      "              >>> record = Entrez.read(handle)\n",
      "              >>> handle.close()\n",
      "\n",
      "           where record is now a Python dictionary or list.\n",
      "\n",
      "        - parse        Parses the XML results returned by those of the above functions\n",
      "          which can return multiple records - such as efetch, esummary\n",
      "          and elink. Typical usage is:\n",
      "\n",
      "              >>> handle = Entrez.esummary(db=\"pubmed\", id=\"19304878,14630660\", retmode=\"xml\")\n",
      "              >>> records = Entrez.parse(handle)\n",
      "              >>> for record in records:\n",
      "              ...     # each record is a Python dictionary or list.\n",
      "              ...     print(record['Title'])\n",
      "              Biopython: freely available Python tools for computational molecular biology and bioinformatics.\n",
      "              PDB file parser and structure class implemented in Python.\n",
      "              >>> handle.close()\n",
      "\n",
      "          This function is appropriate only if the XML file contains\n",
      "          multiple records, and is particular useful for large files.\n",
      "\n",
      "        - _open        Internally used function.\n",
      "\n",
      "PACKAGE CONTENTS\n",
      "    Parser\n",
      "\n",
      "FUNCTIONS\n",
      "    ecitmatch(**keywds)\n",
      "        Retrieve PMIDs for input citation strings, returned as a handle.\n",
      "\n",
      "        ECitMatch retrieves PubMed IDs (PMIDs) that correspond to a set of input\n",
      "        citation strings.\n",
      "\n",
      "        See the online documentation for an explanation of the parameters:\n",
      "        http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.ECitMatch\n",
      "\n",
      "        Short example:\n",
      "\n",
      "        >>> from Bio import Entrez\n",
      "        >>> Entrez.email = \"Your.Name.Here@example.org\"\n",
      "        >>> citation_1 = {\"journal_title\": \"proc natl acad sci u s a\",\n",
      "        ...               \"year\": \"1991\", \"volume\": \"88\", \"first_page\": \"3248\",\n",
      "        ...               \"author_name\": \"mann bj\", \"key\": \"citation_1\"}\n",
      "        >>> handle = Entrez.ecitmatch(db=\"pubmed\", bdata=[citation_1])\n",
      "        >>> print(handle.read().strip().split(\"|\"))\n",
      "        ['proc natl acad sci u s a', '1991', '88', '3248', 'mann bj', 'citation_1', '2014248']\n",
      "        >>> handle.close()\n",
      "\n",
      "        :returns: Handle to the results, by default in plain text.\n",
      "        :raises urllib.error.URLError: If there's a network error.\n",
      "\n",
      "    efetch(db, **keywords)\n",
      "        Fetch Entrez results which are returned as a handle.\n",
      "\n",
      "        EFetch retrieves records in the requested format from a list or set of one or\n",
      "        more UIs or from user's environment.\n",
      "\n",
      "        See the online documentation for an explanation of the parameters:\n",
      "        http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EFetch\n",
      "\n",
      "        Short example:\n",
      "\n",
      "        >>> from Bio import Entrez\n",
      "        >>> Entrez.email = \"Your.Name.Here@example.org\"\n",
      "        >>> handle = Entrez.efetch(db=\"nucleotide\", id=\"AY851612\", rettype=\"gb\", retmode=\"text\")\n",
      "        >>> print(handle.readline().strip())\n",
      "        LOCUS       AY851612                 892 bp    DNA     linear   PLN 10-APR-2007\n",
      "        >>> handle.close()\n",
      "\n",
      "        This will automatically use an HTTP POST rather than HTTP GET if there\n",
      "        are over 200 identifiers as recommended by the NCBI.\n",
      "\n",
      "        **Warning:** The NCBI changed the default retmode in Feb 2012, so many\n",
      "        databases which previously returned text output now give XML.\n",
      "\n",
      "        :returns: Handle to the results.\n",
      "        :raises urllib.error.URLError: If there's a network error.\n",
      "\n",
      "    egquery(**keywds)\n",
      "        Provide Entrez database counts for a global search.\n",
      "\n",
      "        EGQuery provides Entrez database counts in XML for a single search\n",
      "        using Global Query.\n",
      "\n",
      "        See the online documentation for an explanation of the parameters:\n",
      "        http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EGQuery\n",
      "\n",
      "        This quick example based on a longer version from the Biopython\n",
      "        Tutorial just checks there are over 60 matches for 'Biopython'\n",
      "        in PubMedCentral:\n",
      "\n",
      "        >>> from Bio import Entrez\n",
      "        >>> Entrez.email = \"Your.Name.Here@example.org\"\n",
      "        >>> handle = Entrez.egquery(term=\"biopython\")\n",
      "        >>> record = Entrez.read(handle)\n",
      "        >>> handle.close()\n",
      "        >>> for row in record[\"eGQueryResult\"]:\n",
      "        ...     if \"pmc\" in row[\"DbName\"]:\n",
      "        ...         print(int(row[\"Count\"]) > 60)\n",
      "        True\n",
      "\n",
      "        :returns: Handle to the results, by default in XML format.\n",
      "        :raises urllib.error.URLError: If there's a network error.\n",
      "\n",
      "    einfo(**keywds)\n",
      "        Return a summary of the Entrez databases as a results handle.\n",
      "\n",
      "        EInfo provides field names, index term counts, last update, and\n",
      "        available links for each Entrez database.\n",
      "\n",
      "        See the online documentation for an explanation of the parameters:\n",
      "        http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EInfo\n",
      "\n",
      "        Short example:\n",
      "\n",
      "        >>> from Bio import Entrez\n",
      "        >>> Entrez.email = \"Your.Name.Here@example.org\"\n",
      "        >>> record = Entrez.read(Entrez.einfo())\n",
      "        >>> 'pubmed' in record['DbList']\n",
      "        True\n",
      "\n",
      "        :returns: Handle to the results, by default in XML format.\n",
      "        :raises urllib.error.URLError: If there's a network error.\n",
      "\n",
      "    elink(**keywds)\n",
      "        Check for linked external articles and return a handle.\n",
      "\n",
      "        ELink checks for the existence of an external or Related Articles link\n",
      "        from a list of one or more primary IDs;  retrieves IDs and relevancy\n",
      "        scores for links to Entrez databases or Related Articles; creates a\n",
      "        hyperlink to the primary LinkOut provider for a specific ID and\n",
      "        database, or lists LinkOut URLs and attributes for multiple IDs.\n",
      "\n",
      "        See the online documentation for an explanation of the parameters:\n",
      "        http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.ELink\n",
      "\n",
      "        Note that ELink treats the \"id\" parameter differently than the other\n",
      "        tools when multiple values are given. You should generally pass multiple\n",
      "        UIDs as a list of strings or integers. This will provide a \"one-to-one\"\n",
      "        mapping from source database UIDs to destination database UIDs in the\n",
      "        result. If multiple source UIDs are passed as a single comma-delimited\n",
      "        string all destination UIDs will be mixed together in the result.\n",
      "\n",
      "        This example finds articles related to the Biopython application\n",
      "        note's entry in the PubMed database:\n",
      "\n",
      "        >>> from Bio import Entrez\n",
      "        >>> Entrez.email = \"Your.Name.Here@example.org\"\n",
      "        >>> pmid = \"19304878\"\n",
      "        >>> handle = Entrez.elink(dbfrom=\"pubmed\", id=pmid, linkname=\"pubmed_pubmed\")\n",
      "        >>> record = Entrez.read(handle)\n",
      "        >>> handle.close()\n",
      "        >>> print(record[0][\"LinkSetDb\"][0][\"LinkName\"])\n",
      "        pubmed_pubmed\n",
      "        >>> linked = [link[\"Id\"] for link in record[0][\"LinkSetDb\"][0][\"Link\"]]\n",
      "        >>> \"14630660\" in linked\n",
      "        True\n",
      "\n",
      "        This is explained in much more detail in the Biopython Tutorial.\n",
      "\n",
      "        :returns: Handle to the results, by default in XML format.\n",
      "        :raises urllib.error.URLError: If there's a network error.\n",
      "\n",
      "    epost(db, **keywds)\n",
      "        Post a file of identifiers for future use.\n",
      "\n",
      "        Posts a file containing a list of UIs for future use in the user's\n",
      "        environment to use with subsequent search strategies.\n",
      "\n",
      "        See the online documentation for an explanation of the parameters:\n",
      "        http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EPost\n",
      "\n",
      "        :returns: Handle to the results.\n",
      "        :raises urllib.error.URLError: If there's a network error.\n",
      "\n",
      "    esearch(db, term, **keywds)\n",
      "        Run an Entrez search and return a handle to the results.\n",
      "\n",
      "        ESearch searches and retrieves primary IDs (for use in EFetch, ELink\n",
      "        and ESummary) and term translations, and optionally retains results\n",
      "        for future use in the user's environment.\n",
      "\n",
      "        See the online documentation for an explanation of the parameters:\n",
      "        http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.ESearch\n",
      "\n",
      "        Short example:\n",
      "\n",
      "        >>> from Bio import Entrez\n",
      "        >>> Entrez.email = \"Your.Name.Here@example.org\"\n",
      "        >>> handle = Entrez.esearch(\n",
      "        ...     db=\"nucleotide\", retmax=10, idtype=\"acc\",\n",
      "        ...     term=\"opuntia[ORGN] accD 2007[Publication Date]\"\n",
      "        ... )\n",
      "        ...\n",
      "        >>> record = Entrez.read(handle)\n",
      "        >>> handle.close()\n",
      "        >>> int(record[\"Count\"]) >= 2\n",
      "        True\n",
      "        >>> \"EF590893.1\" in record[\"IdList\"]\n",
      "        True\n",
      "        >>> \"EF590892.1\" in record[\"IdList\"]\n",
      "        True\n",
      "\n",
      "        :returns: Handle to the results, which are always in XML format.\n",
      "        :raises urllib.error.URLError: If there's a network error.\n",
      "\n",
      "    espell(**keywds)\n",
      "        Retrieve spelling suggestions as a results handle.\n",
      "\n",
      "        ESpell retrieves spelling suggestions, if available.\n",
      "\n",
      "        See the online documentation for an explanation of the parameters:\n",
      "        http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.ESpell\n",
      "\n",
      "        Short example:\n",
      "\n",
      "        >>> from Bio import Entrez\n",
      "        >>> Entrez.email = \"Your.Name.Here@example.org\"\n",
      "        >>> record = Entrez.read(Entrez.espell(term=\"biopythooon\"))\n",
      "        >>> print(record[\"Query\"])\n",
      "        biopythooon\n",
      "        >>> print(record[\"CorrectedQuery\"])\n",
      "        biopython\n",
      "\n",
      "        :returns: Handle to the results, by default in XML format.\n",
      "        :raises urllib.error.URLError: If there's a network error.\n",
      "\n",
      "    esummary(**keywds)\n",
      "        Retrieve document summaries as a results handle.\n",
      "\n",
      "        ESummary retrieves document summaries from a list of primary IDs or\n",
      "        from the user's environment.\n",
      "\n",
      "        See the online documentation for an explanation of the parameters:\n",
      "        http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.ESummary\n",
      "\n",
      "        This example discovers more about entry 19923 in the structure\n",
      "        database:\n",
      "\n",
      "        >>> from Bio import Entrez\n",
      "        >>> Entrez.email = \"Your.Name.Here@example.org\"\n",
      "        >>> handle = Entrez.esummary(db=\"structure\", id=\"19923\")\n",
      "        >>> record = Entrez.read(handle)\n",
      "        >>> handle.close()\n",
      "        >>> print(record[0][\"Id\"])\n",
      "        19923\n",
      "        >>> print(record[0][\"PdbDescr\"])\n",
      "        Crystal Structure Of E. Coli Aconitase B\n",
      "\n",
      "\n",
      "        :returns: Handle to the results, by default in XML format.\n",
      "        :raises urllib.error.URLError: If there's a network error.\n",
      "\n",
      "    parse(source, validate=True, escape=False, ignore_errors=False)\n",
      "        Parse an XML file from the NCBI Entrez Utilities into python objects.\n",
      "\n",
      "        This function parses an XML file created by NCBI's Entrez Utilities,\n",
      "        returning a multilevel data structure of Python lists and dictionaries.\n",
      "        This function is suitable for XML files that (in Python) can be represented\n",
      "        as a list of individual records. Whereas 'read' reads the complete file\n",
      "        and returns a single Python list, 'parse' is a generator function that\n",
      "        returns the records one by one. This function is therefore particularly\n",
      "        useful for parsing large files.\n",
      "\n",
      "        Most XML files returned by NCBI's Entrez Utilities can be parsed by\n",
      "        this function, provided its DTD is available. Biopython includes the\n",
      "        DTDs for most commonly used Entrez Utilities.\n",
      "\n",
      "        The argument ``source`` must be a file or file-like object opened in binary\n",
      "        mode, or a filename. The parser detects the encoding from the XML file, and\n",
      "        uses it to convert all text in the XML to the correct Unicode string. The\n",
      "        functions in Bio.Entrez to access NCBI Entrez will automatically return XML\n",
      "        data in binary mode. For files, use mode \"rb\" when opening the file, as in\n",
      "\n",
      "            >>> from Bio import Entrez\n",
      "            >>> path = \"Entrez/pubmed1.xml\"\n",
      "            >>> stream = open(path, \"rb\")  # opened in binary mode\n",
      "            >>> records = Entrez.parse(stream)\n",
      "            >>> for record in records:\n",
      "            ...     print(record['MedlineCitation']['Article']['Journal']['Title'])\n",
      "            ...\n",
      "            Social justice (San Francisco, Calif.)\n",
      "            Biochimica et biophysica acta\n",
      "            >>> stream.close()\n",
      "\n",
      "        Alternatively, you can use the filename directly, as in\n",
      "\n",
      "            >>> records = Entrez.parse(path)\n",
      "            >>> for record in records:\n",
      "            ...     print(record['MedlineCitation']['Article']['Journal']['Title'])\n",
      "            ...\n",
      "            Social justice (San Francisco, Calif.)\n",
      "            Biochimica et biophysica acta\n",
      "\n",
      "        which is safer, as the file stream will automatically be closed after all\n",
      "        the records have been read, or if an error occurs.\n",
      "\n",
      "        If validate is True (default), the parser will validate the XML file\n",
      "        against the DTD, and raise an error if the XML file contains tags that\n",
      "        are not represented in the DTD. If validate is False, the parser will\n",
      "        simply skip such tags.\n",
      "\n",
      "        If escape is True, all characters that are not valid HTML are replaced\n",
      "        by HTML escape characters to guarantee that the returned strings are\n",
      "        valid HTML fragments. For example, a less-than sign (<) is replaced by\n",
      "        &lt;. If escape is False (default), the string is returned as is.\n",
      "\n",
      "        If ignore_errors is False (default), any error messages in the XML file\n",
      "        will raise a RuntimeError. If ignore_errors is True, error messages will\n",
      "        be stored as ErrorElement items, without raising an exception.\n",
      "\n",
      "        Whereas the data structure seems to consist of generic Python lists,\n",
      "        dictionaries, strings, and so on, each of these is actually a class\n",
      "        derived from the base type. This allows us to store the attributes\n",
      "        (if any) of each element in a dictionary my_element.attributes, and\n",
      "        the tag name in my_element.tag.\n",
      "\n",
      "    read(source, validate=True, escape=False, ignore_errors=False)\n",
      "        Parse an XML file from the NCBI Entrez Utilities into python objects.\n",
      "\n",
      "        This function parses an XML file created by NCBI's Entrez Utilities,\n",
      "        returning a multilevel data structure of Python lists and dictionaries.\n",
      "        Most XML files returned by NCBI's Entrez Utilities can be parsed by\n",
      "        this function, provided its DTD is available. Biopython includes the\n",
      "        DTDs for most commonly used Entrez Utilities.\n",
      "\n",
      "        The argument ``source`` must be a file or file-like object opened in binary\n",
      "        mode, or a filename. The parser detects the encoding from the XML file, and\n",
      "        uses it to convert all text in the XML to the correct Unicode string. The\n",
      "        functions in Bio.Entrez to access NCBI Entrez will automatically return XML\n",
      "        data in binary mode. For files, use mode \"rb\" when opening the file, as in\n",
      "\n",
      "            >>> from Bio import Entrez\n",
      "            >>> path = \"Entrez/esearch1.xml\"\n",
      "            >>> stream = open(path, \"rb\")  # opened in binary mode\n",
      "            >>> record = Entrez.read(stream)\n",
      "            >>> print(record['QueryTranslation'])\n",
      "            biopython[All Fields]\n",
      "            >>> stream.close()\n",
      "\n",
      "        Alternatively, you can use the filename directly, as in\n",
      "\n",
      "            >>> record = Entrez.read(path)\n",
      "            >>> print(record['QueryTranslation'])\n",
      "            biopython[All Fields]\n",
      "\n",
      "        which is safer, as the file stream will automatically be closed after the\n",
      "        record has been read, or if an error occurs.\n",
      "\n",
      "        If validate is True (default), the parser will validate the XML file\n",
      "        against the DTD, and raise an error if the XML file contains tags that\n",
      "        are not represented in the DTD. If validate is False, the parser will\n",
      "        simply skip such tags.\n",
      "\n",
      "        If escape is True, all characters that are not valid HTML are replaced\n",
      "        by HTML escape characters to guarantee that the returned strings are\n",
      "        valid HTML fragments. For example, a less-than sign (<) is replaced by\n",
      "        &lt;. If escape is False (default), the string is returned as is.\n",
      "\n",
      "        If ignore_errors is False (default), any error messages in the XML file\n",
      "        will raise a RuntimeError. If ignore_errors is True, error messages will\n",
      "        be stored as ErrorElement items, without raising an exception.\n",
      "\n",
      "        Whereas the data structure seems to consist of generic Python lists,\n",
      "        dictionaries, strings, and so on, each of these is actually a class\n",
      "        derived from the base type. This allows us to store the attributes\n",
      "        (if any) of each element in a dictionary my_element.attributes, and\n",
      "        the tag name in my_element.tag.\n",
      "\n",
      "DATA\n",
      "    __warningregistry__ = {'version': 430, ('\\n            Email address i...\n",
      "    api_key = None\n",
      "    email = None\n",
      "    max_tries = 3\n",
      "    sleep_between_tries = 15\n",
      "    tool = 'biopython'\n",
      "\n",
      "FILE\n",
      "    /home/operon/Storage/micromamba/envs/gorilla/lib/python3.12/site-packages/Bio/Entrez/__init__.py\n",
      "\n"
     ]
    }
   ],
   "source": [
    "entrez_docs = pydoc.doc(Bio.Entrez)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Provides code to access NCBI over the WWW.\\n\\nThe main Entrez web page is available at:\\nhttp://www.ncbi.nlm.nih.gov/Entrez/\\n\\nEntrez Programming Utilities web page is available at:\\nhttp://www.ncbi.nlm.nih.gov/books/NBK25501/\\n\\nThis module provides a number of functions like ``efetch`` (short for\\nEntrez Fetch) which will return the data as a handle object. This is\\na standard interface used in Python for reading data from a file, or\\nin this case a remote network connection, and provides methods like\\n``.read()`` or offers iteration over the contents line by line. See\\nalso \"What the heck is a handle?\" in the Biopython Tutorial and\\nCookbook: http://biopython.org/DIST/docs/tutorial/Tutorial.html\\nhttp://biopython.org/DIST/docs/tutorial/Tutorial.pdf\\nThe handle returned by these functions can be either in text mode or\\nin binary mode, depending on the data requested and the results\\nreturned by NCBI Entrez. Typically, XML data will be in binary mode\\nwhile other data will be in text mode, as required by the downstream\\nparser to parse the data.\\n\\nUnlike a handle to a file on disk from the ``open(filename)`` function,\\nwhich has a ``.name`` attribute giving the filename, the handles from\\n``Bio.Entrez`` all have a ``.url`` attribute instead giving the URL\\nused to connect to the NCBI Entrez API.\\n\\nThe ``epost``, ``efetch``, and ``esummary`` tools take an \"id\" parameter\\nwhich corresponds to one or more database UIDs (or accession.version\\nidentifiers in the case of sequence databases such as \"nuccore\" or\\n\"protein\"). The Python value of the \"id\" keyword passed to these functions\\nmay be either a single ID as a string or integer or multiple IDs as an\\niterable of strings/integers. You may also pass a single string containing\\nmultiple IDs delimited by commas. The ``elink`` tool also accepts multiple\\nIDs but the argument is handled differently than the other three. See that\\nfunction\\'s docstring for more information.\\n\\nAll the functions that send requests to the NCBI Entrez API will\\nautomatically respect the NCBI rate limit (of 3 requests per second\\nwithout an API key, or 10 requests per second with an API key) and\\nwill automatically retry when encountering transient failures\\n(i.e. connection failures or HTTP 5XX codes). By default, Biopython\\ndoes a maximum of three tries before giving up, and sleeps for 15\\nseconds between tries. You can tweak these parameters by setting\\n``Bio.Entrez.max_tries`` and ``Bio.Entrez.sleep_between_tries``.\\n\\nThe Entrez module also provides an XML parser which takes a handle\\nas input.\\n\\nVariables:\\n\\n    - email        Set the Entrez email parameter (default is not set).\\n    - tool         Set the Entrez tool parameter (default is ``biopython``).\\n    - api_key      Personal API key from NCBI. If not set, only 3 queries per\\n      second are allowed. 10 queries per seconds otherwise with a\\n      valid API key.\\n    - max_tries    Configures how many times failed requests will be\\n      automatically retried on error (default is 3).\\n    - sleep_between_tries   The delay, in seconds, before retrying a request on\\n      error (default is 15).\\n\\nFunctions:\\n\\n    - efetch       Retrieves records in the requested format from a list of one or\\n      more primary IDs or from the user\\'s environment\\n    - epost        Posts a file containing a list of primary IDs for future use in\\n      the user\\'s environment to use with subsequent search strategies\\n    - esearch      Searches and retrieves primary IDs (for use in EFetch, ELink,\\n      and ESummary) and term translations and optionally retains\\n      results for future use in the user\\'s environment.\\n    - elink        Checks for the existence of an external or Related Articles link\\n      from a list of one or more primary IDs.  Retrieves primary IDs\\n      and relevancy scores for links to Entrez databases or Related\\n      Articles;  creates a hyperlink to the primary LinkOut provider\\n      for a specific ID and database, or lists LinkOut URLs\\n      and Attributes for multiple IDs.\\n    - einfo        Provides field index term counts, last update, and available\\n      links for each database.\\n    - esummary     Retrieves document summaries from a list of primary IDs or from\\n      the user\\'s environment.\\n    - egquery      Provides Entrez database counts in XML for a single search\\n      using Global Query.\\n    - espell       Retrieves spelling suggestions.\\n    - ecitmatch    Retrieves PubMed IDs (PMIDs) that correspond to a set of\\n      input citation strings.\\n\\n    - read         Parses the XML results returned by any of the above functions.\\n      Alternatively, the XML data can be read from a file opened in binary mode.\\n      Typical usage is:\\n\\n          >>> from Bio import Entrez\\n          >>> Entrez.email = \"Your.Name.Here@example.org\"\\n          >>> handle = Entrez.einfo() # or esearch, efetch, ...\\n          >>> record = Entrez.read(handle)\\n          >>> handle.close()\\n\\n       where record is now a Python dictionary or list.\\n\\n    - parse        Parses the XML results returned by those of the above functions\\n      which can return multiple records - such as efetch, esummary\\n      and elink. Typical usage is:\\n\\n          >>> handle = Entrez.esummary(db=\"pubmed\", id=\"19304878,14630660\", retmode=\"xml\")\\n          >>> records = Entrez.parse(handle)\\n          >>> for record in records:\\n          ...     # each record is a Python dictionary or list.\\n          ...     print(record[\\'Title\\'])\\n          Biopython: freely available Python tools for computational molecular biology and bioinformatics.\\n          PDB file parser and structure class implemented in Python.\\n          >>> handle.close()\\n\\n      This function is appropriate only if the XML file contains\\n      multiple records, and is particular useful for large files.\\n\\n    - _open        Internally used function.\\n\\n'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Bio.Entrez.__doc__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "docs = dict()\n",
    "for name, val in Entrez.__dict__.items():\n",
    "    if callable(val) and name[0] == \"e\":\n",
    "        docs[name] = val.__doc__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "epost Post a file of identifiers for future use.\n",
      "\n",
      "    Posts a file containing a list of UIs for future use in the user's\n",
      "    environment to use with subsequent search strategies.\n",
      "\n",
      "    See the online documentation for an explanation of the parameters:\n",
      "    http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EPost\n",
      "\n",
      "    :returns: Handle to the results.\n",
      "    :raises urllib.error.URLError: If there's a network error.\n",
      "    \n",
      "efetch Fetch Entrez results which are returned as a handle.\n",
      "\n",
      "    EFetch retrieves records in the requested format from a list or set of one or\n",
      "    more UIs or from user's environment.\n",
      "\n",
      "    See the online documentation for an explanation of the parameters:\n",
      "    http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EFetch\n",
      "\n",
      "    Short example:\n",
      "\n",
      "    >>> from Bio import Entrez\n",
      "    >>> Entrez.email = \"Your.Name.Here@example.org\"\n",
      "    >>> handle = Entrez.efetch(db=\"nucleotide\", id=\"AY851612\", rettype=\"gb\", retmode=\"text\")\n",
      "    >>> print(handle.readline().strip())\n",
      "    LOCUS       AY851612                 892 bp    DNA     linear   PLN 10-APR-2007\n",
      "    >>> handle.close()\n",
      "\n",
      "    This will automatically use an HTTP POST rather than HTTP GET if there\n",
      "    are over 200 identifiers as recommended by the NCBI.\n",
      "\n",
      "    **Warning:** The NCBI changed the default retmode in Feb 2012, so many\n",
      "    databases which previously returned text output now give XML.\n",
      "\n",
      "    :returns: Handle to the results.\n",
      "    :raises urllib.error.URLError: If there's a network error.\n",
      "    \n",
      "esearch Run an Entrez search and return a handle to the results.\n",
      "\n",
      "    ESearch searches and retrieves primary IDs (for use in EFetch, ELink\n",
      "    and ESummary) and term translations, and optionally retains results\n",
      "    for future use in the user's environment.\n",
      "\n",
      "    See the online documentation for an explanation of the parameters:\n",
      "    http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.ESearch\n",
      "\n",
      "    Short example:\n",
      "\n",
      "    >>> from Bio import Entrez\n",
      "    >>> Entrez.email = \"Your.Name.Here@example.org\"\n",
      "    >>> handle = Entrez.esearch(\n",
      "    ...     db=\"nucleotide\", retmax=10, idtype=\"acc\",\n",
      "    ...     term=\"opuntia[ORGN] accD 2007[Publication Date]\"\n",
      "    ... )\n",
      "    ...\n",
      "    >>> record = Entrez.read(handle)\n",
      "    >>> handle.close()\n",
      "    >>> int(record[\"Count\"]) >= 2\n",
      "    True\n",
      "    >>> \"EF590893.1\" in record[\"IdList\"]\n",
      "    True\n",
      "    >>> \"EF590892.1\" in record[\"IdList\"]\n",
      "    True\n",
      "\n",
      "    :returns: Handle to the results, which are always in XML format.\n",
      "    :raises urllib.error.URLError: If there's a network error.\n",
      "    \n",
      "elink Check for linked external articles and return a handle.\n",
      "\n",
      "    ELink checks for the existence of an external or Related Articles link\n",
      "    from a list of one or more primary IDs;  retrieves IDs and relevancy\n",
      "    scores for links to Entrez databases or Related Articles; creates a\n",
      "    hyperlink to the primary LinkOut provider for a specific ID and\n",
      "    database, or lists LinkOut URLs and attributes for multiple IDs.\n",
      "\n",
      "    See the online documentation for an explanation of the parameters:\n",
      "    http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.ELink\n",
      "\n",
      "    Note that ELink treats the \"id\" parameter differently than the other\n",
      "    tools when multiple values are given. You should generally pass multiple\n",
      "    UIDs as a list of strings or integers. This will provide a \"one-to-one\"\n",
      "    mapping from source database UIDs to destination database UIDs in the\n",
      "    result. If multiple source UIDs are passed as a single comma-delimited\n",
      "    string all destination UIDs will be mixed together in the result.\n",
      "\n",
      "    This example finds articles related to the Biopython application\n",
      "    note's entry in the PubMed database:\n",
      "\n",
      "    >>> from Bio import Entrez\n",
      "    >>> Entrez.email = \"Your.Name.Here@example.org\"\n",
      "    >>> pmid = \"19304878\"\n",
      "    >>> handle = Entrez.elink(dbfrom=\"pubmed\", id=pmid, linkname=\"pubmed_pubmed\")\n",
      "    >>> record = Entrez.read(handle)\n",
      "    >>> handle.close()\n",
      "    >>> print(record[0][\"LinkSetDb\"][0][\"LinkName\"])\n",
      "    pubmed_pubmed\n",
      "    >>> linked = [link[\"Id\"] for link in record[0][\"LinkSetDb\"][0][\"Link\"]]\n",
      "    >>> \"14630660\" in linked\n",
      "    True\n",
      "\n",
      "    This is explained in much more detail in the Biopython Tutorial.\n",
      "\n",
      "    :returns: Handle to the results, by default in XML format.\n",
      "    :raises urllib.error.URLError: If there's a network error.\n",
      "    \n",
      "einfo Return a summary of the Entrez databases as a results handle.\n",
      "\n",
      "    EInfo provides field names, index term counts, last update, and\n",
      "    available links for each Entrez database.\n",
      "\n",
      "    See the online documentation for an explanation of the parameters:\n",
      "    http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EInfo\n",
      "\n",
      "    Short example:\n",
      "\n",
      "    >>> from Bio import Entrez\n",
      "    >>> Entrez.email = \"Your.Name.Here@example.org\"\n",
      "    >>> record = Entrez.read(Entrez.einfo())\n",
      "    >>> 'pubmed' in record['DbList']\n",
      "    True\n",
      "\n",
      "    :returns: Handle to the results, by default in XML format.\n",
      "    :raises urllib.error.URLError: If there's a network error.\n",
      "    \n",
      "esummary Retrieve document summaries as a results handle.\n",
      "\n",
      "    ESummary retrieves document summaries from a list of primary IDs or\n",
      "    from the user's environment.\n",
      "\n",
      "    See the online documentation for an explanation of the parameters:\n",
      "    http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.ESummary\n",
      "\n",
      "    This example discovers more about entry 19923 in the structure\n",
      "    database:\n",
      "\n",
      "    >>> from Bio import Entrez\n",
      "    >>> Entrez.email = \"Your.Name.Here@example.org\"\n",
      "    >>> handle = Entrez.esummary(db=\"structure\", id=\"19923\")\n",
      "    >>> record = Entrez.read(handle)\n",
      "    >>> handle.close()\n",
      "    >>> print(record[0][\"Id\"])\n",
      "    19923\n",
      "    >>> print(record[0][\"PdbDescr\"])\n",
      "    Crystal Structure Of E. Coli Aconitase B\n",
      "\n",
      "\n",
      "    :returns: Handle to the results, by default in XML format.\n",
      "    :raises urllib.error.URLError: If there's a network error.\n",
      "    \n",
      "egquery Provide Entrez database counts for a global search.\n",
      "\n",
      "    EGQuery provides Entrez database counts in XML for a single search\n",
      "    using Global Query.\n",
      "\n",
      "    See the online documentation for an explanation of the parameters:\n",
      "    http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EGQuery\n",
      "\n",
      "    This quick example based on a longer version from the Biopython\n",
      "    Tutorial just checks there are over 60 matches for 'Biopython'\n",
      "    in PubMedCentral:\n",
      "\n",
      "    >>> from Bio import Entrez\n",
      "    >>> Entrez.email = \"Your.Name.Here@example.org\"\n",
      "    >>> handle = Entrez.egquery(term=\"biopython\")\n",
      "    >>> record = Entrez.read(handle)\n",
      "    >>> handle.close()\n",
      "    >>> for row in record[\"eGQueryResult\"]:\n",
      "    ...     if \"pmc\" in row[\"DbName\"]:\n",
      "    ...         print(int(row[\"Count\"]) > 60)\n",
      "    True\n",
      "\n",
      "    :returns: Handle to the results, by default in XML format.\n",
      "    :raises urllib.error.URLError: If there's a network error.\n",
      "    \n",
      "espell Retrieve spelling suggestions as a results handle.\n",
      "\n",
      "    ESpell retrieves spelling suggestions, if available.\n",
      "\n",
      "    See the online documentation for an explanation of the parameters:\n",
      "    http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.ESpell\n",
      "\n",
      "    Short example:\n",
      "\n",
      "    >>> from Bio import Entrez\n",
      "    >>> Entrez.email = \"Your.Name.Here@example.org\"\n",
      "    >>> record = Entrez.read(Entrez.espell(term=\"biopythooon\"))\n",
      "    >>> print(record[\"Query\"])\n",
      "    biopythooon\n",
      "    >>> print(record[\"CorrectedQuery\"])\n",
      "    biopython\n",
      "\n",
      "    :returns: Handle to the results, by default in XML format.\n",
      "    :raises urllib.error.URLError: If there's a network error.\n",
      "    \n",
      "ecitmatch Retrieve PMIDs for input citation strings, returned as a handle.\n",
      "\n",
      "    ECitMatch retrieves PubMed IDs (PMIDs) that correspond to a set of input\n",
      "    citation strings.\n",
      "\n",
      "    See the online documentation for an explanation of the parameters:\n",
      "    http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.ECitMatch\n",
      "\n",
      "    Short example:\n",
      "\n",
      "    >>> from Bio import Entrez\n",
      "    >>> Entrez.email = \"Your.Name.Here@example.org\"\n",
      "    >>> citation_1 = {\"journal_title\": \"proc natl acad sci u s a\",\n",
      "    ...               \"year\": \"1991\", \"volume\": \"88\", \"first_page\": \"3248\",\n",
      "    ...               \"author_name\": \"mann bj\", \"key\": \"citation_1\"}\n",
      "    >>> handle = Entrez.ecitmatch(db=\"pubmed\", bdata=[citation_1])\n",
      "    >>> print(handle.read().strip().split(\"|\"))\n",
      "    ['proc natl acad sci u s a', '1991', '88', '3248', 'mann bj', 'citation_1', '2014248']\n",
      "    >>> handle.close()\n",
      "\n",
      "    :returns: Handle to the results, by default in plain text.\n",
      "    :raises urllib.error.URLError: If there's a network error.\n",
      "    \n"
     ]
    }
   ],
   "source": [
    "for name, val in docs.items():\n",
    "    print(name, val)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/operon/Storage/micromamba/envs/gorilla/lib/python3.12/site-packages/Bio/Entrez/__init__.py:723: UserWarning: \n",
      "            Email address is not specified.\n",
      "\n",
      "            To make use of NCBI's E-utilities, NCBI requires you to specify your\n",
      "            email address with each request.  As an example, if your email address\n",
      "            is A.N.Other@example.com, you can specify it as follows:\n",
      "               from Bio import Entrez\n",
      "               Entrez.email = 'A.N.Other@example.com'\n",
      "            In case of excessive usage of the E-utilities, NCBI will attempt to contact\n",
      "            a user at the email address provided before blocking access to the\n",
      "            E-utilities.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Count': '42', 'RetMax': '20', 'RetStart': '0', 'IdList': ['5609316', '5187760', '5134066', '5116506', '5078804', '5078798', '5049814', '4476448', '4456634', '4399472', '4398940', '3414666', '3295634', '3295110', '3295106', '3069802', '3069790', '3012282', '3012280', '3012268'], 'TranslationSet': [], 'TranslationStack': [{'Term': 'Entrez[All Fields]', 'Field': 'All Fields', 'Count': '1181', 'Explode': 'N'}, {'Term': 'Programming[All Fields]', 'Field': 'All Fields', 'Count': '5609', 'Explode': 'N'}, 'AND', 'GROUP'], 'QueryTranslation': 'Entrez[All Fields] AND Programming[All Fields]'}\n"
     ]
    }
   ],
   "source": [
    "with Entrez.esearch(db=\"books\", term='Entrez Programming') as handle:\n",
    "    record = Entrez.read(handle)\n",
    "print(record)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "ids = \",\".join(record[\"IdList\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/operon/Storage/micromamba/envs/gorilla/lib/python3.12/site-packages/Bio/Entrez/__init__.py:723: UserWarning: \n",
      "            Email address is not specified.\n",
      "\n",
      "            To make use of NCBI's E-utilities, NCBI requires you to specify your\n",
      "            email address with each request.  As an example, if your email address\n",
      "            is A.N.Other@example.com, you can specify it as follows:\n",
      "               from Bio import Entrez\n",
      "               Entrez.email = 'A.N.Other@example.com'\n",
      "            In case of excessive usage of the E-utilities, NCBI will attempt to contact\n",
      "            a user at the email address provided before blocking access to the\n",
      "            E-utilities.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "with Entrez.efetch(db=\"books\", id=ids) as handle:\n",
    "    record = Entrez.read(handle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['5609316', '5187760', '5134066', '5116506', '5078804', '5078798', '5049814', '4476448', '4456634', '4399472', '4398940', '3414666', '3295634', '3295110', '3295106', '3069802', '3069790', '3012282', '3012280', '3012268']"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "record"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
